---
title: "Preprocessing Short-Read Sequencing Files for Isoformic"
date: 2025-10-02
date-modified: 2025-11-11
date-format: long
author:
  - name: Lucio Rezende Queiroz
  - name: Izabela Mamede Conceição
vignette: >
  %\VignetteIndexEntry{Preprocessing Sequencing Files}
  %\VignetteEngine{quarto::html}
  %\VignetteEncoding{UTF-8}
knitr:
  opts_chunk:
    collapse: true
    comment: "#>"
---

```{r}
#| label: setup
#| include: false
knitr::opts_chunk$set(echo = TRUE)
```

##  Salmon: command command line interface anatomy

### From the Terminal - Salmon Command Line Interface

Below is a single-sample invocation (to be run in a Bash shell) showing the
typical arguments we used when making the mini dataset used in this vignette.

Replace the paths with yours if you want to reproduce.

More thorough information can be found on: <https://combine-lab.github.io/salmon/>.

``` bash
# Not needed as of Salmon 1.10.3:
  # --validateMappings \               # selective-alignment (more specific, recommended)

salmon quant \
  --libType A \                      # A -> autodetect library type from the reads
  --index /path/to/index \           # Transcriptome index (here: index built from the FASTA annotation file)
  --mates1 /path/SRR..._1.fastq \    # R1 FASTQ
  --mates2 /path/SRR..._2.fastq \    # R2 FASTQ
  --output /path/SRR..._quant \      # Output folder (contains quant.sf, aux_info/)
  --threads 8 \                      # Number of parallel threads used
  --disableChainingHeuristic \  # improve mapping sensitivity, but slower
  --allowDovetail \               # allow dovetailed alignments
  --softclipOverhangs \          # softclip reads that overhang transcripts start and end
  --dumpEq \
  --dumpEqWeights \               # dump equivalent class information, needed by Terminus
  --posBias --seqBias --gcBias \     # bias corrections (positional, sequence, GC)
  --numGibbsSamples 100              # draw inferential replicates; required for fishpond/Swish
```

### From R - `condathis` wrapper for Salmon

Caveats: currently [Salmon](https://anaconda.org/bioconda/salmon) is only available for macOS and Linux on the [Bioconda](https://bioconda.github.io/) channel.

Therefore, this example will not work natively on Windows.
But you can use WSL2 (Windows Subsystem for Linux) on Windows 10/11 to run it.

```{r}
#| eval: false
library(isoformic)
tx_fasta_path <- download_reference(version = "49", file_type = "fasta")
# TODO: @luciorq Use genome fasta to add decoy sequences
# genome_fasta_path <- download_reference(version = "49", file_type = "genome_fasta")
gff_path <- download_reference(version = "49", file_type = "gff")

base_dir <- fs::path("data-raw", "vignette_data")
# fs::dir_create(base_dir)
# TODO: @luciorq trim reads? `fastp -i SRRxxxx.fastq.gz -o SRRxxxx_trimmed.fastq.gz -q 10 -w 8`

# condathis::get_env_dir("salmon-env")
# micromamba run -p /Users/luciorq/.local/share/R/condathis/envs/salmon-env salmon index --help
# micromamba run -p /Users/luciorq/.local/share/R/condathis/envs/salmon-env salmon quant --help-reads

index_path <- fs::path(base_dir, "salmon_index_k15")
salmon_index(
  fasta_path = tx_fasta_path,
  index_path = index_path,
  kmer_len = 15,
  num_threads = 8,
  is_gencode = TRUE # to handle transcript names properly,
  # decoy_fasta = genome_fasta_path
)

fastq_files <- fs::dir_ls("data-raw/mini_data/sub_fastq/")

sample_names <- fs::path_file(fastq_files) |>
  stringr::str_remove("_sub_[1-2].fastq.gz$") |>
  unique() |>
  sort()

for (i in seq_along(sample_names)) {
  sample_name <- sample_names[i]

  salmon_output <- fs::path(base_dir, "salmon_quants", paste0(sample_name, "_quant"))
  fastq_r1 <- fs::dir_ls("data-raw/mini_data/sub_fastq/", regexp = paste0(sample_name, "_sub_1.fastq.gz$"))
  fastq_r2 <- fs::dir_ls("data-raw/mini_data/sub_fastq/", regexp = paste0(sample_name, "_sub_2.fastq.gz$"))
  # fs::file_exists(fastq_r1)
  # fs::file_exists(fastq_r2)

  salmon_quant(
    index_path = index_path,
    input_r1 = fastq_r1,
    input_r2 = fastq_r2,
    output_dir = salmon_output,
    num_threads = 8,
    num_gibbs = 20,
    min_score_fraction = "0.30"
  )
}
```

Now to run and get the isoformic tables.

```{r}
#| eval: false
base_dir <- fs::path("data-raw", "vignette_data")
quants_dir <- fs::path(base_dir, "salmon_quants")
sample_dirs <- list.dirs(quants_dir, full.names = FALSE, recursive = FALSE)
sample_names <- sub("_quant$", "", sample_dirs)
quant_files <- file.path(quants_dir, paste0(sample_names, "_quant"), "quant.sf")
names(quant_files) <- sample_names
condition <- c(rep("control", 3), rep("treatment", 3))

col_data <- tibble::tibble(
  files = quant_files,
  names = sample_names,
  condition = factor(condition, levels = c("control", "treatment"))
)
col_data

txi_tx <- tximport::tximport(
  files = col_data$files,
  type = "salmon",
  txIn = TRUE,
  txOut = TRUE,
  countsFromAbundance = "no",
  dropInfReps = FALSE,
  ignoreAfterBar = TRUE
)


# lobstr::obj_sizes(txi_tx)
#> 636.22 MB

class(txi_tx$infReps$SRR11498039)
# tx_ids_to_keep <- rownames(se_tx)
tx_ids_to_keep <- c("ENST00000367468.10", "ENST00000681605.1", "ENST00000680451.1",
"ENST00000490885.6", "ENST00000559627.1", "ENST00000466691.1",
"ENST00000263341.7", "ENST00000491056.5", "ENST00000487639.1",
"ENST00000418817.5", "ENST00000432018.5", "ENST00000416750.1",
"ENST00000496280.5", "ENST00000477398.1", "ENST00000673816.1",
"ENST00000673847.1", "ENST00000415035.2", "ENST00000698141.1",
"ENST00000698142.1", "ENST00000698143.1", "ENST00000361099.8",
"ENST00000673762.1", "ENST00000540176.6", "ENST00000698144.1",
"ENST00000698145.1", "ENST00000698146.1", "ENST00000698147.1",
"ENST00000698148.1", "ENST00000698149.1", "ENST00000674081.1",
"ENST00000673832.1", "ENST00000673863.1", "ENST00000423282.2",
"ENST00000673858.1", "ENST00000674028.1", "ENST00000673777.1",
"ENST00000673942.1", "ENST00000673734.1", "ENST00000409465.5",
"ENST00000674080.1", "ENST00000698150.1", "ENST00000673952.1",
"ENST00000674153.1", "ENST00000392322.7", "ENST00000673841.1",
"ENST00000452281.6", "ENST00000673885.1", "ENST00000392323.6",
"ENST00000464072.1", "ENST00000673859.1", "ENST00000698151.1",
"ENST00000424722.6", "ENST00000673638.1", "ENST00000454414.5",
"ENST00000432058.1", "ENST00000696131.1", "ENST00000401931.2",
"ENST00000696132.1", "ENST00000483500.1", "ENST00000307407.8",
"ENST00000306602.3", "ENST00000449264.3", "ENST00000699334.1",
"ENST00000404625.5", "ENST00000426291.5", "ENST00000401651.5",
"ENST00000464710.2", "ENST00000485300.1", "ENST00000258743.10",
"ENST00000407492.5", "ENST00000401630.7", "ENST00000406575.1",
"ENST00000697954.1", "ENST00000697955.1", "ENST00000697956.1",
"ENST00000697957.1", "ENST00000553342.2", "ENST00000216797.10",
"ENST00000554001.5", "ENST00000697958.1", "ENST00000557140.5",
"ENST00000697959.1", "ENST00000557389.1", "ENST00000697960.1",
"ENST00000697961.1", "ENST00000557459.2", "ENST00000697965.1",
"ENST00000555371.1", "ENST00000697962.1", "ENST00000697966.1",
"ENST00000557100.5", "ENST00000556664.1", "ENST00000555629.1",
"ENST00000580907.6", "ENST00000225831.4", "ENST00000624362.2",
"ENST00000605140.6", "ENST00000651122.1", "ENST00000605509.2",
"ENST00000330871.3", "ENST00000587578.1", "ENST00000423829.2",
"ENST00000588645.1", "ENST00000264832.8", "ENST00000585443.1",
"ENST00000592686.1", "ENST00000613606.2", "ENST00000618639.4",
"ENST00000632752.1", "ENST00000445232.2", "ENST00000376122.3"
)

# se_tx <- SummarizedExperiment(
#   assays = list(
#     counts = txi_tx$counts,
#     abundance = txi_tx$abundance,
#     length = txi_tx$length
#   ),
#   colData = col_data,
#   metadata = list(
#     txi = list(
#       countsFromAbundance = txi_tx$countsFromAbundance
#     )
#   )
# )

se_tx_all <- tximeta::tximeta(
  coldata = col_data,
  type = "salmon",
  txOut = TRUE,
  skipMeta = TRUE,
  useHub = FALSE,
  ignoreAfterBar = TRUE,
  skipSeqinfo = TRUE
)
se_tx_with_inf <- se_tx_all[tx_ids_to_keep, ]
inf_assay_list <- assays(se_tx_with_inf)[4:23]
inf_assay_list |>
  readr::write_rds("inst/extdata/mini_inf_reps.rds", compress = "xz")
```

## Actual Vignette Starts Here

```{r}
suppressPackageStartupMessages({
  library(tximeta)
  library(SummarizedExperiment)
  library(fishpond)
  library(DESeq2)
})

library(isoformic)

# Where extdata lives
extdata_dir <- system.file("extdata", package = "isoformic")

quants_dir <- file.path(extdata_dir, "mini_quants")

gff_path <- download_reference(version = "49", file_type = "gff")
```

## Load Salmon Quantification with tximport

Point to Salmon's `quant.sf` output files.

```{r}
# discover sample folders like SRR11498039_quant, etc.
sample_dirs <- list.dirs(quants_dir, full.names = FALSE, recursive = FALSE)

# keep only *_quant and strip the suffix for sample names
sample_names <- sub("_quant$", "", sample_dirs)
stopifnot(length(sample_names) == 6)

# define groups: "control" = first 3, "treatment" = last 3 (sorted by sample ID)
sample_names <- sort(sample_names)
condition <- c(rep("control", 3), rep("treatment", 3))

quant_files <- file.path(quants_dir, paste0(sample_names, "_quant"), "quant.sf")
names(quant_files) <- sample_names
stopifnot(all(file.exists(quant_files)))

col_data <- tibble::tibble(
  files = quant_files,
  names = sample_names,
  condition = factor(condition, levels = c("control", "treatment"))
)
col_data
```

Import transcriptomics data with tximport.

```{r}
se_tx <- tximeta::tximeta(
  coldata = col_data,
  type = "salmon",
  txOut = TRUE,
  skipMeta = TRUE,
  useHub = FALSE,
  ignoreAfterBar = TRUE,
  skipSeqinfo = TRUE
)

mini_inf_reps <- readRDS(system.file("extdata/mini_inf_reps.rds", package = "isoformic"))
assays(se_tx) <- c(assays(se_tx), mini_inf_reps)
```

Convert to IsoformicExperiment object.

```{r}
iso <- as_isoformic(se_tx, gff_path)
iso
```

```{r}
tx_to_gene_df <- tx_to_gene(iso)
head(tx_to_gene_df)

# rowData(se_tx) <- row_data(iso)
iso
```

Perform transcript-level Differential Expression Analysis with fishpond.

```{r}
y <- se_tx
y <- fishpond::scaleInfReps(y)

y <- fishpond::labelKeep(y, minCount = 10, minN = 3)
# y <- y[mcols(y)$keep, ]

colData(y)$condition <- factor(col_data$condition, levels = c("control", "treatment"))
# y$condition <- factor(y$condition, levels = c("control", "treatment"))
colData(y)

# run Swish
y <- fishpond::swish(y, x = "condition")

dea_tx_res <- S4Vectors::mcols(y) |>
  base::as.data.frame() |>
  tibble::rownames_to_column("transcript_id") |>
  tibble::as_tibble()

iso@dea$det <- dea_tx_res

de_tx(iso)
```

Summarize Transcript SummarizedExperiment to Gene level.

```{r}
se_gene <- tximeta::summarizeToGene(
  se_tx,
  tx2gene = dplyr::select(row_data(iso), transcript_id, gene_id),
  skipRanges = TRUE
)
```

Perform Gene level Differential Expression with DESeq2.

```{r}
colData(se_gene)$condition <- factor(colData(se_gene)$condition, levels = c("control", "treatment"))

assays(se_gene)$counts <- round(assays(se_gene)$counts)

dds <- DESeq2::DESeqDataSet(se_gene, design = ~condition)
dds <- dds[rowSums(DESeq2::counts(dds) >= 10) >= 3, ]
dds <- DESeq2::DESeq(dds)

res <- DESeq2::results(dds, contrast = c("condition", "treatment", "control"))

# shrink for stability on small example
if (rlang::is_installed("apeglm")) {
  message("apeglm is installed")
  shrink_type <- "apeglm"
} else {
  message("apeglm is NOT installed")
  shrink_type <- "normal"
}

res_shr <- DESeq2::lfcShrink(
  dds,
  coef = "condition_treatment_vs_control",
  res = res,
  type = shrink_type
)

# res
# res_shr
dea_gene_res <- base::as.data.frame(res_shr) |>
  tibble::rownames_to_column("gene_id") |>
  tibble::as_tibble() |>
  dplyr::rename(
    log2FC = log2FoldChange,
    pvalue = pvalue,
    qvalue = padj
  )


iso@dea$deg <- dea_gene_res

de_gene(iso)
```

Plot examples.

```{r}
top_de_gene <- de_gene(iso) |>
  dplyr::arrange(qvalue) |>
  dplyr::slice_head(n = 1) |>
  dplyr::pull(gene_id)

plot_log2fc(
  iso,
  feature = top_de_gene,
  feature_column = "gene_id"
)
```

By gene name

```{r}
gene_to_plot <- row_data(iso) |>
  dplyr::filter(gene_id %in% top_de_gene) |>
  dplyr::slice_head(n = 1) |>
  dplyr::pull(gene_name)

plot_log2fc(
  iso,
  feature = gene_to_plot,
  feature_column = "gene_name"
)
```

Top DE transcripts

```{r}
row_data(iso) |>
  dplyr::filter(gene_name %in% "MTG2")

plot_log2fc(
  iso,
  feature = "MTG2",
  feature_column = "gene_name"
)
```
