---
title: "Isoformic:  Isoform-level biological interpretation of transcriptomic data - Non-coding RNA"
date: "2023-02-09"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true
author:
  - name: Izabela Mamede Conceição
    affiliation: Laboratory of Genetics Biochemistry, Universidade Federal de Minas Gerais, Belo Horizonte, Brazil
  - name: Lucio Rezende Queiroz
    affiliation: Department of Pathology and Laboratory Medicine, Weill Cornell Medicine
vignette: |
  %\VignetteIndexEntry{intro-isoformic}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: sentence
---

### Intro

```{r}
library(tidyverse)
```


One type of RNA that we did not touch on the first Isoformic Vignette is the long non-coding RNAs.
This second Vignette will focus on their analysis and replicate what our group did on <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9607547/>.

One of the things to note that for this analysis the depth of your transcriptome will matter even more if compared to the isoform analysis for productive genes.
Even for higher depth transcriptomes the cutoff values for the differential expression of lncRNA isoforms need to be reduced due to the fact these transcripts have, biologically, a lower expression if compared with coding genes.
This makes the sequencing results for them look very distinct based it's depth, if you did not use specific sequencing methods only for lncRNAs.

```{r}
knitr::include_graphics("https://i.imgur.com/I0nXu5U.png")
```

This will consist of two main concepts: 1.The analysis of differentially expressed genes on the region around a selected lncRNA. 2. The enrichment of all targets correlated to another selected lncRNA.
On the first point we will be looking for the cis-action of lncRNAs, when they are transcribed and act on the same locus and on the second point we will detect pathways in which the mRNAs correlated to a certain lncRNA can be acting.
The cis-action is more well-confirmable and documentable with only bioinformatic analysis but the trans action will be performed in a guilty-by-association approach using correlation with other transcript's expression.

## TADs and cis-action

TADs or Topologically Associating Domains are genomic domains structured on the three-dimensional genome organization.
TADs vary from cell to cell and in treatment conditions.
Their border contributes to the genomic expression inside that region, restricting interactions between promoters, enhancers and their target sequences.
Recently lncRNAs present in those borders have risen as important regulators of the genes inside each TAD (<https://link.springer.com/article/10.1186/s13059-018-1405-5>) this local action of the RNAs is what we will explore on the first steps of the pipeline.

The main input for this pipeline will be our DET_final table witch was outputted from the first part of the other pipeline, we will use it to select all lncRNA isoforms in our data. We will also do a soft filtering of pvalue < 0.05 and absolute log2FC higher then 0.5. This is a good time to also filter a bit our lncRNA list for the specific targets we want to analize since if not the correlation analysis will become a very computationally demanding task.

```{r}
lnc_for_analysis <- c("UCA1-238","AC008687.3-201",
                      "DUBR-230","DUBR-202","DUBR-213",
                      "DUBR-227", "LINC01095-203",
                      "PART1-203", "LCAL1-202",
                      "NAV2-AS4-201", "AL139383.1-201",
                      "NRAD1-202", "NRAD1-205",
                      "LINC00551-203", "LINC02291-201",
                      "EWSAT1-201", "AC124248.2-201",
                      "C1QTNF1-AS1-202",
                      "AP005212.4-201",
                      "ZNF667-AS1-211", "XIST-229",
                      "XIST-223")
```

```{r}
PE1_lncRNAs <-PE1_DETs_final %>% filter(transcript_type == "lncRNA" & abs(log2FC)> 0.5 & pvalue < 0.05 & transcript_name %in% lnc_for_analysis)

head(PE1_lncRNAs)
```

So our first goal will be to get every transcript that is located around our lncRNAs. TAD lengths vary widely but over 80% of all TADs fall below 2 megabases, so here we will use 1 megabase to each side as a cut for possible cis-action.

The information of genomic coordinates will be done using the .gff we had downloaded and the package GenomicFeatures.

```{r, message=FALSE, warning=FALSE}
library(GenomicFeatures)
txdbv34 <- makeTxDbFromGFF("data-raw/gencode.v34.chr_patch_hapl_scaff.annotation.gff3.gz")
txdbv34
```


We will make a vector using the transcript ids of the lncRNAs from our table
```{r}
keys <-PE1_lncRNAs$transcript_id
```

Now we use biomaRt select function to select the transcript start, transcript end transcript strand and chromosome columns for our selected lncRNA. You can reduce even more this list if you have a smaller number of lncRNAs you want to analize. Reducing this number is good for running the transcript correlation later on. The time it takes for it will increase exponentially base don the ammount of lncRNA you have here.

```{r}
columns(txdbv34)

cols_i_want <- c("TXSTART", "TXEND", "TXSTRAND", "TXCHROM")

lncRNAisos_start_and_end <-biomaRt::select(txdbv34, keys=keys, columns=cols_i_want, keytype="TXNAME")

```

With those ranges we will make new ones to be used for filtering. Increasing the start and end columns by 1.000.000. You just need to pass the table we had on the last step and then the value you want to add to each side of the lncRNA position.

```{r}
selected_lnc_1mb <-modify_lnc_ranges(lncRNAisos_start_and_end, 1000000)
```

Using this we can now reverse engineer Granges to give us all the transcripts inside those coordinates.

```{r}
selected_granges_1mb <-  GenomicRanges::makeGRangesFromDataFrame(selected_lnc_1mb,
                         keep.extra.columns=TRUE,
                         ignore.strand=FALSE,
                         seqinfo=NULL,
                         seqnames.field=c("TXCHROM"),
                         start.field="new_start",
                         end.field="new_end",
                         strand.field="TXSTRAND",
                         starts.in.df.are.0based=FALSE)

head(selected_granges_1mb)
```

Now we just use GRanges framework to count for us the overlaps between two objects: first the ranges of 2mb around our selected lncRNA and second the names of the transcripts in our txdb, that are the same names which were in the differential expression and the other Isoformic analyses.

```{r}
transcripts_all <-transcripts(txdbv34)

number_overlaps <-GenomicRanges::countOverlaps(selected_granges_1mb, transcripts_all)
```

With that we can plot a barplot of how many other transcripts appear on the genomic regions around each lncRNA

```{r}
lnc_names <-selected_lnc_1mb %>%
  left_join(tx_to_gene, by = c("TXNAME"="transcript_id")) 
```

```{r, fig.height=3, fig.width=5}
barplot(number_overlaps, names.arg = lnc_names$transcript_name,
        main="Possible targets in 1mb",ylab="Count",las=2)
```

We can also use the function subsetByOverlaps to detect exactly which RNA, inside our annotation we want according to the range filtering.

```{r}
subset_lnc_targets <-as.data.frame(subsetByOverlaps(transcripts_all,selected_granges_1mb))

subset_lnc_targets
```

Now we have around 2000 transcripts that are in the region around our lncRNAs. To get which transcripts are these we just join them to the tx2gene table

```{r}
subset_lnc_targets_res  <-left_join(subset_lnc_targets, tx_to_gene, by = c("tx_name"="transcript_id"))
```

### Correlation analysis

All the steps for the correlation will be done using the counts table, we will filter it first using our selected lncRNA, and then again by the transcripts in the vicinity table


First the counts only for the lncRNA
```{r}
counts_lnc_names <-PE1_counts %>% filter(transcript_id %in% selected_lnc_1mb$TXNAME)

counts_lnc <-counts_lnc_names %>% dplyr::select(-transcript_name, -transcript_id)

rownames(counts_lnc) <- counts_lnc_names$transcript_name

head(counts_lnc)
```

Now the counts for the selected targets

```{r}
counts_targets_names <-PE1_counts %>% filter(transcript_name %in% subset_lnc_targets_res$transcript_name)
counts_targets <-counts_targets_names %>% dplyr::select(-transcript_name, -transcript_id)

rownames(counts_targets) <- counts_targets_names$transcript_name

head(counts_targets) 
```


For the test itself we will use the rcorr function from Hmisc package that computes correlation and pvalue  together. We will also transpose the matrices so our transcript_name information now corresponds to the columns.

```{r}
cor_lnc_targets <-Hmisc::rcorr(t(counts_lnc), t(counts_targets), type = "spearman")
```

Now we have a matrix with the spearman correlations and the respective pvalue for each of every transcript on the vicinity of every lncRNA in our list. To turn that into a tidy data-frama with those results we will use the flattenCorrMatrix function, this function was extracted from: https://github.com/ameya225/useful_functions/blob/master/flattenCorrMatrix.R.

```{r}
cor_lnc_targets_P <-cor_lnc_targets$P

cor_lnc_targets_r<-cor_lnc_targets$r

flattenCorrMatrix <- function(cormat, pmat) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut],
    p = pmat[ut]
  )
}
cor_lnc_targets_res  <- flattenCorrMatrix(cor_lnc_targets_r, cor_lnc_targets_P)
```

One problem with our table is that there are some correlations of the transcripts with itself and some correlations not between lncRNA and a gene transcript. So we do the following filtering
```{r}
cor_lnc_targets_final <-cor_lnc_targets_res %>% filter(row != column) %>%
  filter(row %in% lnc_for_analysis | column %in% lnc_for_analysis)

```

So in the end we have around 16.000 correlations from all the mRNAs, many of them not significant.

```{r}
PE1_DETs_final_filt <-PE1_DETs_final %>% filter(pvalue < 0.05 & abs(log2FC)> 1)

cor_lnc_targets_final_filt <-cor_lnc_targets_final %>% filter(row %in% PE1_DETs_final_filt$transcript_name & column  %in% PE1_DETs_final_filt$transcript_name)
```


Now we need to go back to our TAD information. Which, of those correlated mRNAs, are in the same genomic region as our lncRNA isoforms?

First we use our tx_to_gene table to ass the ensembl ID information
```{r}
cor_lnc_targets_final_filt <- cor_lnc_targets_final_filt %>% left_join(tx_to_gene, by = c("row"="transcript_name")) %>% dplyr::select(1:5)

cor_lnc_targets_final_filt <- cor_lnc_targets_final_filt %>% left_join(tx_to_gene, by = c("column"="transcript_name")) %>% dplyr::select(1:6)

names(cor_lnc_targets_final_filt) <- c("row", "column", "cor", "p", "transcript_id_row", "transcript_id_col")

```


We can now use the subsetbyOverlaps function and the GRanges objcts we had done to gte a data.frame of all transcripts on our ranges
```{r}
all_transcripts_onranges <-as.data.frame(subsetByOverlaps(transcripts_all, selected_granges_1mb))
```

We need to add the chromossome information on our corr table
```{r}
transcripts_all_s <-as.data.frame(transcripts_all) %>% dplyr::select(seqnames, tx_name)

joined_row <-cor_lnc_targets_final_filt %>% left_join(as.data.frame(transcripts_all_s), by = c("transcript_id_row"="tx_name"))

joined_all <-joined_row %>% left_join(as.data.frame(transcripts_all_s), by = c("transcript_id_col"="tx_name"))

```

And then filter our correlations on that ranges table
```{r}
correlations_and_ranges <-joined_all %>% filter(transcript_id_row %in% all_transcripts_onranges$tx_name & seqnames.x==seqnames.y) %>% dplyr::select(-5,-6,-7,-8)
```

### Visualization

With that table ready we can then add the FoldChange values and select correlations of transcripts from different genes

```{r}
correlations_and_ranges_dic <-correlations_and_ranges %>% left_join(PE1_DETs_final, by = c("row"="transcript_name"))


correlations_and_ranges_dic <-correlations_and_ranges_dic %>% left_join(PE1_DETs_final, by = c("column"="transcript_name"))

correlations_and_ranges_dic_filt <-correlations_and_ranges_dic %>% filter(gene_name.x != gene_name.y & transcript_type.x != "lncRNA")
```

We will prepare the table to make a heatmap of the mRNA lncRNA isoform pairs.

```{r}
lnc_cor_forplot <-correlations_and_ranges_dic_filt %>% dplyr::select(row, column, log2FC.x, log2FC.y, transcript_type.x)

lnc_cor_forplot$pair <- c()

lnc_cor_forplot$pair <-paste0(lnc_cor_forplot$row,"_",lnc_cor_forplot$column)


lnc_cor_heatmap<-lnc_cor_forplot %>% dplyr::select(3,4)

rownames(lnc_cor_heatmap) <- lnc_cor_forplot$pair

anno_heatmap <-lnc_cor_forplot %>% dplyr::select(transcript_type.x)

rownames(anno_heatmap) <- lnc_cor_forplot$pair

names(lnc_cor_heatmap) <- c("log2FC_mRNA", "log2FC_lncRNA")

```

And plot the heatmap
```{r, fig.width=5}
g <-pheatmap::pheatmap(lnc_cor_heatmap, annotation_row = anno_heatmap)
```

The duos found here can later be visualized on a platform like UCSC Genome Browser.


## Trans-action

At this point we already have almost all the information we need  for the inference of pathways associated with possible trans-acting targets.
What we will do here is wuite direct, we will use the correlation value os the lncRNA-mRNA pairs to enrich and with that we will have Normalized Enrichment Scores which vary depending of the expression direction of the lncRNA and of it`s target.

As an example we will choose one lncRNAs isoform and enrich it`s targets.

```{r}
"LINC01095-203"

```










