---
title: "Isoformic Application to Mouse Neuron Dataset"
date: "2024-04-25"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true
author:
  - name: Izabela Mamede Conceição
    affiliation: Laboratory of Genetics Biochemistry, Universidade Federal de Minas Gerais, Belo Horizonte, Brazil
  - name: Lucio Rezende Queiroz
    affiliation: Department of Pathology and Laboratory Medicine, Weill Cornell Medicine, New York, USA
  - name: Carlos Augusto da Mata-Machado
    affiliation: Laboratory of Genetics Biochemistry, Universidade Federal de Minas Gerais, Belo Horizonte, Brazil
vignette: |
  %\VignetteIndexEntry{isoformic-example}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

## Information

Isoformic uses the differential expression data derived from the transcript counts and Gibbs samples created by pseudo-alignment software such as Salmon (Patro et al., 2017) along with the GENCODE annotation to explore biological function at the isoform-level. Part 1 of this article gives a demonstration of the pipeline upstream of the functions covered in the [introduction vignette](https://luciorq.github.io/isoformic/articles/isoformic_intro.html), such as importing Salmon's output, Differential Expression and saving DET, DEG and TPM tables; further, in this document's [part 2](#part-2---isoformic-application), we cover Isoformic's functions. The dataset we chose for this demonstration was [PRJNA984605 BioProject](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA984605), which examined chronic stress in mice (Degroat et al., 2024).

------------------------------------------------------------------------

# Part 1 - Importing and Differential Expression {#part-1---importing-and-differential-expression}

Load dependency packages:

```{r, echo = T, results = 'hide', message = FALSE, warning=FALSE}
library(isoformic)
library(tidyverse)
library(tximport)
library(DESeq2)
library(apeglm)
library(fishpond)
library(readr)
library(tximeta)
```

## Importing files

Before importing Salmon's output, we load a sample table which contains sample names and associated metadata.

```{r}
index_anxiety <- readr::read_csv("ref/index_anxiety.csv",
                                 show_col_types = FALSE)
head(index_anxiety)
```

The sample names are then used to assign a vector with the Salmon output (`quant.sf` files) locations.

```{r}
files <- file.path("quant_files/application_quant", paste0(index_anxiety$name,"_quant"), "quant.sf")

names(files) <- index_anxiety$name

all(file.exists(files))
```

We also set up a transcript-to-gene "dictionary" table (i.e. transcripts information and their parent genes correspondence) thought Isoformic function `make_tx_to_gene()`, using a FASTA file obtained from the GENCODE website - in this example, *Mus musculus* release v34.

```{r}
txtogene_Mv34 <- isoformic::make_tx_to_gene("ref/gencode.vM34.transcripts.fa.gz", file_type = "fasta")

txtogene_Mv34 <- txtogene_Mv34 |>
  dplyr::select(1,2,5,6,8)

head(txtogene_Mv34)
```

## Importing Salmon's output

Then, `tximport` (Soneson et al., 2015) is used to import Salmon's output (file paths stored in the `files` vector). We save the TPM table from the new object. First at transcript-level:

```{r, message = FALSE}
tximporteddata <- tximport::tximport(files, type="salmon", 
                                     txIn = TRUE, 
                                     txOut = TRUE, 
                                     countsFromAbundance = "no"
                                     )

txi_counts <- as.data.frame(tximporteddata$abundance) |>
  rownames_to_column(var="transcript_id")

write_csv(txi_counts, "processed_tables/DET_anxiety_counts.csv")
```

And then summarized to gene-level:

```{r, message = FALSE}
tx2gene_filt <- txtogene_Mv34 |> 
  select(1,4)

tximporteddata_gene <- tximport::tximport(files, type="salmon", 
                                          txIn = TRUE, 
                                          txOut = FALSE, 
                                          countsFromAbundance = "no", 
                                          tx2gene = tx2gene_filt
                                          )

dxi_counts <- as.data.frame(tximporteddata_gene$abundance)

write_csv(dxi_counts, "processed_tables/DEGanxiety_counts.csv")
```

## Differential expression at gene-level (DEG table)

DESeq2 (Love et al., 2014) is used to perform differential expression at the gene-level.

```{r, message = FALSE, warning=FALSE}
ddsTxi_gene <- DESeq2::DESeqDataSetFromTximport(tximporteddata_gene,
                                                colData = index_anxiety,
                                                design = ~ cond
                                                )

dds_gene <- DESeq2::DESeq(ddsTxi_gene)
```

```{r}
DESeq2::resultsNames(dds_gene)
```

```{r, message = FALSE, warning=FALSE}
res_dds_gene <- DESeq2::results(dds_gene)

res_shr <- DESeq2::lfcShrink(dds_gene, 
                     coef = "cond_stressed_vs_control", 
                     type="apeglm"
                     )
```

And to obtain the DEG table:

```{r}
Male_difexpr <- as.data.frame(res_shr)

Male_difexpr_results <- tibble::rownames_to_column(Male_difexpr)

write.csv(Male_difexpr_results, "processed_tables/Male_DEG_results_dic.csv")
```

## Differential expression at transcript-level (DET)

### Wrangling

Before the differential expression, we wrangle the sample table to include files paths:

```{r}
files_df <- as.data.frame(files)

files_df <- tibble::rownames_to_column(files_df)

files_df <- dplyr::left_join(files_df, 
                      index_anxiety, 
                      by = c("rowname"="name")
                      )
```

And then filter for male:

```{r, message = FALSE}
files_male <- files_df |> 
  dplyr::filter(sex != "female")

files_male_df <- files_male |>
  dplyr::select(1,2,4)

names(files_male_df) <- c("names", "files", "condition")

se_male <- tximeta::tximeta(files_male_df)

se_male$condition <- factor(se_male$condition, levels = c("control", "stressed"))

y_male <- se_male
```

### Differential expression

To perform transcript-level abundance estimation, we highly recommend the `swish` method implemented in the `fishpond` package (Zhu et al., 2019). In this next code chunk, In this next step, the inferential replicates are scaled, and the RangedSummarizedExperiment is subset to retain only features with ≥100 counts in at least 3 samples.

```{r}
y_male <- fishpond::scaleInfReps(y_male)

y_male <- fishpond::labelKeep(y_male, minCount = 100, minN = 3)

y_male <- y_male[mcols(y_male)$keep,]

y_male <- fishpond::swish(y_male, x = "condition")
```

To save the DET table, we extract the differential expression data from this RangedSummarizedExperiment object.

```{r}
library(S4Vectors)
library(tibble)
library(dplyr)

DET_results_male <- S4Vectors::mcols(y_male) |>
  as.data.frame() |>
  dplyr::mutate(gene_id = sapply(gene_id, `[`, 1)) |>
  tibble::rownames_to_column("transcript_id") |>
  tibble::as_tibble()

DET_results_male_f <- dplyr::left_join(
  DET_results_male, 
  (txtogene_Mv34 |> 
     dplyr::select(transcript_id, gene_name, transcript_type, transcript_name)), by = "transcript_id") |> 
  dplyr::select(-tx_name)

head(DET_results_male_f)

write_csv(DET_results_male_f, "processed_tables/swish_DET_male.csv")
```

------------------------------------------------------------------------

# Part 2 - Isoformic application

In this section we provide a demonstration of Isoformic (thoroughly explained in the [introduction vignette](https://luciorq.github.io/isoformic/articles/isoformic_intro.html)), applied to [PRJNA984605 BioProject](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA984605), which examined chronic stress in mice (Degroat et al., 2024). The upstream data wrangling is described in [part 1](#part-1---importing-and-differential-expression).

Load dependency packages:

```{r, message = FALSE}
library(isoformic)
library(tidyverse)
library(tximport)
library(DESeq2)
library(fishpond)
library(readr)
library(tximeta)
library(fs)
library(readr)
library(dplyr)
library(tidyr)
library(stringr)
library(ghibli)
```

## Importing files

Before using Isoformic's functions, we need to properly set up the data:

### DEG table

After importing the differentially expressed genes (DEGs) table, we merge it with a subset of the dictionary object (`txtogene_Mv34`). We rename columns and create a filtered vector, `DEGs_male_new_names_sig`, to keep only the `gene_name` of features with $|$`log2FC`$|$ $\geq 1$ and `pvalue` $\leq 0.05$.

```{r, message = FALSE}
Male_difexpr_results <- readr::read_csv("processed_tables/Male_difexpr_results_dic.csv",
                                        show_col_types = FALSE)

gene_join <- txtogene_Mv34 |>
  dplyr::select(gene_id, gene_name) |>
  dplyr::distinct()

DEGs_male <- Male_difexpr_results |>
  dplyr::left_join(gene_join, by = c("rowname"="gene_name"))

names(DEGs_male)[2] <- "gene_name"
names(DEGs_male)[4] <- "log2FC"

DEGs_male_new_names_sig <- DEGs_male |>
  dplyr::filter(abs(log2FC) >= 1) |>
  dplyr::filter(pvalue <= 0.05) |>
  dplyr::select(gene_name) |>
  tidyr::drop_na()

DEGs_sig_joined <- DEGs_male_new_names_sig |>
  dplyr::left_join(txtogene_Mv34, by = "gene_name")

```

### TPM table

We import the TPM table, too, and we use the dictionary table (`txtogene_Mv34`) to include a `transcript_name` column.

```{r}
txi_counts <- readr::read_csv("processed_tables/DET_anxiety_counts.csv",
                              show_col_types = FALSE)

tpm_join <- txtogene_Mv34 |>
  dplyr::select(transcript_id, transcript_name) |>
  dplyr::distinct()

male_counts <- txi_counts |>
  dplyr::left_join(tpm_join, by = c("transcript_id"))
```

### DET table

Lastly, we read in the differentially expressed transcripts (DETs) table, and filter it to identify significant transcripts based on arbitrary thresholds, $|$`log2FC`$|$ $\geq 2$ and `pvalue` $\leq 0.05$.

```{r}
DET_results_male_f <- readr::read_csv("processed_tables/swish_DET_male.csv",
                                      show_col_types = FALSE)

DETs_male_sig <- DET_results_male_f |> 
  dplyr::filter(pvalue < 0.05 & abs(log2FC)>1)
```

## Colors

The following code chunk is used to assign a color to each type of transcript use throughout the pipeline.

```{r}
fixed_tx_biotypes <- c(
  "gene", "protein_coding", "retained_intron","protein_coding_CDS_not_defined", 
  "nonsense_mediated_decay","lncRNA", "processed_pseudogene", "transcribed_unprocessed_pseudogene",
  "unprocessed_pseudogene", "non_stop_decay", "transcribed_unitary_pseudogene", "pseudogene",
  "unitary_pseudogene"
  )

tx_type_color_names <- c(
  "#fb8072", "#a6d854", "#8da0cb", "#fc8d62",
  "#66c2a5", "#e78ac3", "#ffd92f", "#e5c494",
  "#d9d9d9", "#d9d9d9", "#d9d9d9", "#ffffb3",
  "#d9d9d9"
  )

names(tx_type_color_names) <- fixed_tx_biotypes

tx_type_color_names
```

## Log2FC Plot

This plot compares the log2FC of case vs. control from a gene and its transcripts. To make this kind of plot, we first create a single table that combines gene-level and transcript-level differential expression results, using the Isoformic function `join_DEG_DET()`.

```{r}
DET_results_male_f_sig <- is_deg_sig(DEGs_sig_joined$transcript_name, DET_results_male_f)

DEG_DET_table <- join_DEG_DET(DEGs_male,
                              DET_results_male_f_sig, 
                              logfc_cut = 1, 
                              pval_cut = 0.05)
head(DEG_DET_table)
```

Now it is possible to use the function `plot_log2FC()` for any gene in the table. We provide some examples:

```{r, fig.width=8, fig.height=4}
Trip12log2fc <- plot_log2FC(DEG_DET_table, "Trip12") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
Trip12log2fc
```

```{r, fig.width=12, fig.height=4}
Meg3log2fc <- plot_log2FC(DEG_DET_table, "Meg3") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) 
Meg3log2fc
```

```{r, fig.width=8, fig.height=4}
Map4log2fc <- plot_log2FC(DEG_DET_table, "Map4") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) 
Map4log2fc
```

```{r, include=FALSE}
# Trip12log2fc
# cowplot::ggsave2("plots/log2fc_trip12.pdf", height = 4, width = 6)
# 
# Meg3log2fc
# cowplot::ggsave2("plots/log2fc_meg3.pdf", height = 4, width = 12)
# 
# Map4log2fc
# cowplot::ggsave2("plots/log2fc_map4.pdf", height = 4, width = 6)
```

## Functional Enrichment

To perform functional enrichment, we expand a known `.gmt` for transcript information and enrich each transcript type separately. We chose the M2 gene set from MSigDB and loaded it using `fgsea::gmtPathways()`.

```{r}
genesets_list <- fgsea::gmtPathways(
  gmt.file = "ref/m2.all.v2023.2.Mm.symbols.gmt.txt"
)

head(str(genesets_list[1:5]))
```

The function `run_enrichment()` can read in `genesets_list` (the `.gmt` loaded as a list of each biological process) along with the previously created DET table (`DET_results_male_f`), the dictionary table (`txtogene_Mv34`) and an arbitrary p-value cutoff (i.e. a threshold applied to the resulting enrichment table).

```{r, message=FALSE, warning=FALSE}
enrichment_df <- run_enrichment(
  det_df = DET_results_male_f,
  genesets_list = genesets_list,
  tx_to_gene = txtogene_Mv34,
  pval_cutoff = 0.05
)

head(enrichment_df)
```

We can plot the resulting enrichment for the productive transcripts:

```{r, fig.width = 12, fig.height = 6}
enrichment_df |>
  dplyr::filter((experiment %in% c("protein_coding")) & (abs(NES) > 2)) |>
  dplyr::arrange(padj) |>
  dplyr::slice_head(n = 15) |>
  ggplot2::ggplot(ggplot2::aes(pathway, experiment)) +
  ggplot2::geom_point(ggplot2::aes(col = NES, size = size)) +
  ggplot2::coord_flip() +
  ggplot2::theme_minimal() +
  ghibli::scale_colour_ghibli_c("MononokeMedium")
```

```{r, include=FALSE}
# cowplot::ggsave2("plots/lolipop_enrichment_coding.pdf", width = 12, height = 6)
```

And we can plot, too, the enrichment for the unproductive transcript types:

```{r, fig.width = 12, fig.height = 6}
enrichment_df |>
  dplyr::filter((!experiment %in% c("protein_coding", "unproductive")) & (NES) > 1.7) |>
  dplyr::arrange(padj) |>
  dplyr::slice_head(n = 30) |>
  ggplot2::ggplot(ggplot2::aes(pathway, experiment)) +
  ggplot2::geom_point(ggplot2::aes(col = NES, size = size)) +
  ggplot2::coord_flip() +
  ggplot2::theme_minimal() +
  ghibli::scale_colour_ghibli_c("LaputaMedium")
```

```{r, include=FALSE}
# cowplot::ggsave2("plots/lolipop_enrichment_RI.pdf", width = 8, height = 4)
```

------------------------------------------------------------------------

<!-- Fill barplot -->

```{r, include=FALSE}

# DETs_male_sig |>
#   ggplot() +
#   geom_bar(aes(x=fct_infreq(transcript_type), fill = transcript_type))+
#   theme_bw()+
#   scale_fill_manual(values = tx_type_color_names)+
#   theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
#   
# 
# cowplot::ggsave2("plots/barplot_counts_type.pdf", width = 6, height = 4)

```

# Genomic Context Plot

To visualize and compare transcripts of a given gene, we use the use the `download_reference()` function to download feature annotation and input it in `prepare_exon_annotation()`, along with the gene of interest. Finally, `plot_tx_context()` can be used to create this plot.

```{r, fig.width = 11, fig.height = 22, message=FALSE, warning=FALSE}
library(isoformic)

download_reference(version = "M34", file_type = "gff", organism ="mouse")

exon_df <- prepare_exon_annotation(
  gene_name = "Trip12",
  file_path = "data-raw/gencode.vM34.annotation.gff3.gz",
  file_type = "gff"
)

tx_to_gene <- txtogene_Mv34

exon_df |>
  plot_tx_context()
```

```{r, include=FALSE}
# cowplot::ggsave2("plots/Genomiczcontext_trip12.pdf", width = 11, height = 22)
```

------------------------------------------------------------------------

# Session Information

```{r}
sessionInfo()
```

------------------------------------------------------------------------

# References

Degroat, T. J., Wiersielis, K., Denney, K., Kodali, S., Daisey, S., Tollkuhn, J., Samuels, B. A., & Roepke, T. A. (2023). Chronic stress and its effects on behavior, RNA expression of the bed nucleus of the stria terminalis, and the M-current of NPY neurons. *Psychoneuroendocrinology*, *161*, 106920. <https://doi.org/10.1016/j.psyneuen.2023.106920>

Love, M. I., Huber, W., & Anders, S. (2014). Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. *Genome Biology*, *15*(12). <https://doi.org/10.1186/s13059-014-0550-8>

Patro, R., Duggal, G., Love, M. I., Irizarry, R. A., & Kingsford, C. (2017). Salmon provides fast and bias-aware quantification of transcript expression. *Nature Methods*, *14*(4), 417–419. <https://doi.org/10.1038/nmeth.4197>

Soneson, C., Love, M. I., & Robinson, M. D. (2015). Differential analyses for RNA-seq: transcript-level estimates improve gene-level inferences. *F1000Research*, *4*, 1521. <https://doi.org/10.12688/f1000research.7563.1>

Zhu, A., Srivastava, A., Ibrahim, J. G., Patro, R., & Love, M. I. (2019). Nonparametric expression analysis using inferential replicate counts. *Nucleic Acids Research*, *47*(18), e105. <https://doi.org/10.1093/nar/gkz622>
