<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Isoformic: Isoform-level biological interpretation of transcriptomic data • isoformic</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Isoformic: Isoform-level biological interpretation of transcriptomic data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">isoformic</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/isoformic_intro.html">Isoformic: Isoform-level biological interpretation of transcriptomic data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/luciorq/isoformic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Isoformic: Isoform-level biological interpretation of transcriptomic data</h1>
                        <h4 data-toc-skip class="author">Izabela Mamede
Conceição</h4>
            <address class="author_afil">
      Laboratory of Genetics Biochemistry, Universidade Federal de Minas
Gerais, Belo Horizonte,
Brazil<br><h4 data-toc-skip class="author">Lucio
Rezende Queiroz</h4>
            <address class="author_afil">
      Department of Pathology and Laboratory Medicine, Weill Cornell
Medicine, New York, USA<br><h4 data-toc-skip class="date">2023-04-07</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/luciorq/isoformic/blob/HEAD/vignettes/isoformic_intro.Rmd" class="external-link"><code>vignettes/isoformic_intro.Rmd</code></a></small>
      <div class="d-none name"><code>isoformic_intro.Rmd</code></div>
    </address>
</address>
</div>

    
    
<div class="section level2">
<h2 id="the-isoformic-method">The Isoformic method<a class="anchor" aria-label="anchor" href="#the-isoformic-method"></a>
</h2>
<p>Welcome to <code>isoformic</code>, a workflow for isoform-level
biological interpretation of transcriptomic data.</p>
<div class="section level3">
<h3 id="information">Information<a class="anchor" aria-label="anchor" href="#information"></a>
</h3>
<p>All data used for the examples here were extracted from the paper: <a href="https://pubmed.ncbi.nlm.nih.gov/32306769/" class="external-link">“Landscape of
Dysregulated Placental RNA Editing Associated With Preeclampsia”</a> in
which were analyzed generated RNA-Seq datasets from preeclamptic and
control placentas. This dataset was chosen due for having many
biological replicates with high sequencing depth.</p>
<!--
This tutorial will encompass X different steps divided into X unities for which type of analysis you want to perform.

## Dependencies

> **NOTE**: As of 2024-06-08 The version of the Workflow described in this vignette is not maintained.
  For installing `v0.0.1` of the package and run this workflow use:


``` r
# if (requireNamespace(remotes, quietly = TRUE)) {
#   remotes::install_github("luciorq/isoformic@v0.0.1")
# }
```
-->
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># devtools::load_all()</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/luciorq/isoformic" class="external-link">isoformic</a></span><span class="op">)</span></span></code></pre></div>
<p>Load dependency packages</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org" class="external-link">tidyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="setting-up-the-data">Setting up the data<a class="anchor" aria-label="anchor" href="#setting-up-the-data"></a>
</h3>
<div class="section level4">
<h4 id="initial-considerations">Initial Considerations<a class="anchor" aria-label="anchor" href="#initial-considerations"></a>
</h4>
<p>We highly recommend the use of <a href="https://github.com/COMBINE-lab/salmon" class="external-link">Salmon</a> for
transcript-level abundance estimation and the <a href="https://bioconductor.org/packages/release/bioc/vignettes/fishpond/inst/doc/swish.html" class="external-link"><code>swish</code></a>
method implemented in the <a href="https://bioconductor.org/packages/release/bioc/html/fishpond.html" class="external-link">fishpond</a>
R package for isoform-level differential expression.</p>
<!--
For a general overview of that process follow the tutorial on ["Processing Bulk RNA-Seq data"](./processing-rna-seq-data.Rmd).
-->
<p>The <a href="https://www.gencodegenes.org/human/" class="external-link">GENCODE</a> project
offers a good annotation of isoforms for the human and mouse genomes,
including isoforms of non-coding genes. Using other sources of
annotation can render completely different results for the following
analysis.</p>
</div>
<div class="section level4">
<h4 id="part-1-data-input">Part 1: Data input<a class="anchor" aria-label="anchor" href="#part-1-data-input"></a>
</h4>
<p>In this version you would need two essential input and two optional
inputs.</p>
<p><strong>Essential input 1:</strong> A transcript differential
expression table. This table can be outputted from any kind of
differential expression software you use but it needs to contain 1)
information on Transcript name OR Ensembl Transcript ID per line 2)
log2FoldChange information in a column named “log2FoldChange”. 3)
<code>p-value</code> information in a column named “pvalue”. Any other
columns on the table will not be used on the main analysis. This DET
table should be the UNFILTERED version of your table.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">/private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic</span></span></span></code></pre>
<p>Example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PE1_DETs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span>, <span class="st">"DETs_fixed.csv"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PE1_DETs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 4</span></span></span>
<span><span class="co">##   transcript_id      log2FC  pvalue qvalue</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> ENST00000456328.2 -<span style="color: #BB0000;">0.692</span>  0.019<span style="text-decoration: underline;">9</span>  0.094<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ENST00000488147.1  0.073<span style="text-decoration: underline;">1</span> 0.652   0.812 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> ENST00000466430.5 -<span style="color: #BB0000;">0.146</span>  0.364   0.590 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> ENST00000442987.3 -<span style="color: #BB0000;">0.183</span>  0.084<span style="text-decoration: underline;">1</span>  0.242 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ENST00000494149.2 -<span style="color: #BB0000;">0.728</span>  0.005<span style="text-decoration: underline;">99</span> 0.042<span style="text-decoration: underline;">0</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> ENST00000595919.1 -<span style="color: #BB0000;">0.078</span><span style="color: #BB0000; text-decoration: underline;">9</span> 0.951   0.977</span></span></code></pre>
<p><strong>Essential input 2:</strong> a FASTA file from GENCODE which
corresponds to the same fasta you used for the transcriptome alignment.
These need to be on the same version since here you will use names from
that GENCODE version to do most of the mergings. The annotation used
also need to be GENCODE since it posses the <code>transcript_type</code>
column that will be used as information as well. If you wish you CAN
provide that information through and external source not being GENCODE,
and from that you will need a table with at least three columns: 1)a
<code>transcript_name</code> column that needs to match those gene names
on your DET table and on your TPM table, 2) a gene-name column to tell
which gene those transcripts belong to and 3) a transcript_type
column.</p>
<p>This table will also have to use transcript names from GENCODE and
have a column stating the transcript biotype that is also got from
GENCODE annotation. The statistics can be ‘pvalue’, ‘svalue’ or ‘qvalue’
but a ‘log2FoldChange’ between your case and control conditions is also
needed for most of the plots.</p>
<p><strong>Optional inputs:</strong> 1) A Transcript per million (TPM)
table matching the transcripts in the differential expression table 2) a
GFF3 file of the transcriptome version which corresponds to your FASTA
and 3) a table of differentially expressed genes of that data to also
use as comparison.</p>
<hr>
<p>Any differential expression table can be used here but the pipeline
authors, after multiple testing, reached the conclusion that the swish
implementation for a differential transcript expression analysis using
inferential replicate counts from Salmon is the one that performs the
best for medium to high depth transcriptome libraries when looking at
number of transcripts and significant values.</p>
<p>Not well annotated transcriptomes will not output results as these
and all the tests here mentioned were done using human data.</p>
<hr>
<p>Isoformic makes available example files that can be used as model for
formatting the necessary files for running the workflow. The files can
be found in the following path.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #0000BB; font-weight: bold;">/private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata</span></span></span></code></pre>
<p>And the files available can be seen with:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/DEGs_PE_fixed2.csv</span></span>
<span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/DETs_fixed.csv</span></span>
<span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/PE_1_DETs.csv</span></span>
<span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/PE_1_counts.csv</span></span>
<span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/c2.cp.reactome.v2023.1.Hs.symbols.gmt.txt</span></span>
<span><span class="co">## <span style="color: #BB0000; font-weight: bold;">/private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/example_gencode.v34.annotation.gff.gz</span></span></span>
<span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/example_gencode.v34.annotation.parquet</span></span>
<span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/hsapiens_REAC_subset.gmt</span></span>
<span><span class="co">## <span style="color: #BB00BB; font-weight: bold;">/private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/isoformic-logo.png</span></span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PE1_DETs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span>, <span class="st">"DETs_fixed.csv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">PE1_DEGs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span>, <span class="st">"DEGs_PE_fixed2.csv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">PE1_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span>, <span class="st">"PE_1_counts.csv"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_csv</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>transcript_id <span class="op">=</span> <span class="va">`...1`</span><span class="op">)</span></span></code></pre></div>
<p>Here we load the table which points for the libraries that represent
our cases (treatment) and our controls. In this library, cases are the
pregnant woman with Preeclampsia and controls matched pregnant without
Preeclampsia.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">PE1_counts</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">PE1_counts</span><span class="op">)</span><span class="op">]</span>,</span>
<span>  condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"treatment"</span>, <span class="fl">8</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">PE1_counts</span><span class="op">)</span> <span class="op">-</span> <span class="fl">9</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sample_table</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       samples condition</span></span>
<span><span class="co">## 1 SRR11498039 treatment</span></span>
<span><span class="co">## 2 SRR11498040 treatment</span></span>
<span><span class="co">## 3 SRR11498041 treatment</span></span>
<span><span class="co">## 4 SRR11498042 treatment</span></span>
<span><span class="co">## 5 SRR11498043 treatment</span></span>
<span><span class="co">## 6 SRR11498044 treatment</span></span></code></pre>
<div class="section level5">
<h5 id="download-reference-files">Download reference files<a class="anchor" aria-label="anchor" href="#download-reference-files"></a>
</h5>
<p>The references used for this project were obtained from the <a href="https://www.gencodegenes.org/human/" class="external-link">GENCODE Project</a> version
34 for the Human genome annotation. The annotation file in GFF3 format
was obtained from <a href="https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.chr_patch_hapl_scaff.annotation.gff3.gz" class="external-link uri">https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_34/gencode.v34.chr_patch_hapl_scaff.annotation.gff3.gz</a>.</p>
<p>This step may take a while depending on the speed of your internet
connection.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/download_reference.html">download_reference</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"34"</span>, file_type <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span><span class="co"># download_reference(version = "34", file_type = "gtf")</span></span>
<span><span class="fu"><a href="../reference/download_reference.html">download_reference</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"34"</span>, file_type <span class="op">=</span> <span class="st">"gff"</span><span class="op">)</span></span></code></pre></div>
<p>To download mouse references, it is necessary to include the letter
‘M’ in the version string (e.g., “M37”).</p>
<!--

### Part 2: Create the `linkedTxome` object from GENCODE annotation

`tximeta` is capable of importing most of the annotations automatically,
but for getting the transcript types we will need to create the `linkedTxome` object manually.

The full documentation can be found at the [`tximeta` ](https://bioconductor.org/packages/devel/bioc/vignettes/tximeta/inst/doc/tximeta.html#What_if_checksum_isn%E2%80%99t_known)



``` r
# gtf_file_path <- "data-raw/gencode.v33.chr_patch_hapl_scaff.annotation.gtf.gz"
# fs::file_exists(gtf_file_path)
#
# base_dir <- "data-raw/gencode_v33"
# json_file_path <- paste0(base_dir, ".json")
# fs::dir_create(base_dir)
# tximeta::makeLinkedTxome(
#   indexDir = base_dir,
#   source = "LocalGENCODE",
#   organism = "Homo sapiens",
#   release = "33",
#   genome = "GRCh38",
#   fasta = "https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_33/gencode.v33.transcripts.fa.gz",
#   gtf = "https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_33/gencode.v33.chr_patch_hapl_scaff.annotation.gtf.gz",
#   write = TRUE,
#   jsonFile = json_file_path
# )
#
# tximeta::loadLinkedTxome("data-raw/gencode_v33.json")
```

-->
<hr>
</div>
</div>
<div class="section level4">
<h4 id="part-2-transcript-to-gene-and-gene-to-transcript-reference-tables">Part 2: Transcript to Gene and Gene to transcript reference
tables<a class="anchor" aria-label="anchor" href="#part-2-transcript-to-gene-and-gene-to-transcript-reference-tables"></a>
</h4>
<p>Using the FASTA file from GENCODE we will construct a transcript per
gene dictionary table and add that information to the main DEG, DET and
TPM table. This step will depend a lot on the names of the columns on
your tables so in the Vignette we decided to change names to keep
consistency.</p>
<p>The input used here is a FASTA file containing the transcript
sequences and their annotation information downloaded from the GENCODE
website with the specific version you used for the alignment. In the
case here GENCODE v34.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fasta_path</span> <span class="op">&lt;-</span> <span class="st">"data-raw/gencode.v34.transcripts.fa.gz"</span></span>
<span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_lines.html" class="external-link">read_lines</a></span><span class="op">(</span><span class="va">fasta_path</span>, n_max <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "&gt;ENST00000456328.2|ENSG00000223972.5|OTTHUMG00000000961.2|OTTHUMT00000362751.1|DDX11L1-202|DDX11L1|1657|processed_transcript|"</span></span>
<span><span class="co">## [2] "GTTAACTTGCCGTCAGCCTTTTCTTTGACCTCTTCTTTCTGTTCATGTGTATTTGCTGTC"                                                                 </span></span>
<span><span class="co">## [3] "TCTTAGCCCAGACTTCCCGTGTCCTTTCCACCGGGCCTTTGAGAGGTCACAGGGTCTTGA"                                                                 </span></span>
<span><span class="co">## [4] "TGCTGTGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTG"                                                                 </span></span>
<span><span class="co">## [5] "CAAGCTGAGCACTGGAGTGGAGTTTTCCTGTGGAGAGGAGCCATGCCTAGAGTGGGATGG"</span></span></code></pre>
<p>As that header shows the imported table is still very weird and not
tidy, so we pass it through the <code>make_tx_to_gene</code> function
that will make it tidy and ready for further use.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_to_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_tx_to_gene.html">make_tx_to_gene</a></span><span class="op">(</span></span>
<span>  file_path <span class="op">=</span> <span class="va">fasta_path</span>,</span>
<span>  file_type <span class="op">=</span> <span class="st">"fasta"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx_to_gene</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">##   transcript_id    gene_id havanna_gene_id havanna_transcript_id transcript_name</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> ENST00000456328… ENSG00… OTTHUMG0000000… OTTHUMT00000362751.1  DDX11L1-202    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ENST00000450305… ENSG00… OTTHUMG0000000… OTTHUMT00000002844.2  DDX11L1-201    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> ENST00000488147… ENSG00… OTTHUMG0000000… OTTHUMT00000002839.1  WASH7P-201     </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> ENST00000619216… ENSG00… -               -                     MIR6859-1-201  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ENST00000473358… ENSG00… OTTHUMG0000000… OTTHUMT00000002840.1  MIR1302-2HG-202</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> ENST00000469289… ENSG00… OTTHUMG0000000… OTTHUMT00000002841.2  MIR1302-2HG-201</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 3 more variables: gene_name &lt;chr&gt;, tx_length &lt;chr&gt;, transcript_type &lt;chr&gt;</span></span></span></code></pre>
<p>Now our <code>tx_to_gene</code> table has 6 columns that are in
order: Ensembl transcript id, Ensembl gene id, Havanna gene id, Havanna
transcript id, transcript name, gene name, transcript length and
transcript type. For the DEG, DET and TPM table we will need the Ensembl
gene id, the Gene name and the transcript type information so we can
convert our tables for transcript_name and add the type information and
if the gene is a DE to the DET table.</p>
<p>Select the columns with the gene id and the gene name info</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_to_gene</span> <span class="op">&lt;-</span> <span class="va">tx_to_gene</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="va">transcript_id</span>, <span class="va">gene_id</span>,</span>
<span>    <span class="va">transcript_name</span>, <span class="va">gene_name</span>,</span>
<span>    <span class="va">transcript_type</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p>First we add the <code>gene_name</code> information to the DEG
table</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_join</span> <span class="op">&lt;-</span> <span class="va">tx_to_gene</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_id</span>, <span class="va">gene_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">PE1_DEGs</span> <span class="op">&lt;-</span> <span class="va">PE1_DEGs</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">gene_join</span>, by <span class="op">=</span> <span class="st">"gene_id"</span><span class="op">)</span></span></code></pre></div>
<p>Now the transcript name for the TPM table</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tpm_join</span> <span class="op">&lt;-</span> <span class="va">tx_to_gene</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">transcript_id</span>, <span class="va">transcript_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">PE1_counts</span> <span class="op">&lt;-</span> <span class="va">PE1_counts</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">tpm_join</span>, by <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"transcript_id"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The DET table will be our main table for analysis.</p>
<hr>
</div>
<div class="section level4">
<h4 id="part-3-constructing-the-main-table">Part 3: Constructing the main table<a class="anchor" aria-label="anchor" href="#part-3-constructing-the-main-table"></a>
</h4>
<p>The Gene-level information will input for us categorical values to be
added on the DET table. In more detail: we need to now if that
transcript’s gene (1) pass on the gene-level expression cutoff values
and (2) which type does that transcript belongs to.</p>
<p>There are multiple types on the Ensembl library and some of their
definitions superpose to one another, the ones further analyzed here can
be seen on this figure</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"https://i.imgur.com/UWoAr0k.png"</span><span class="op">)</span></span></code></pre></div>
<p><img src="https://i.imgur.com/UWoAr0k.png"><!-- --></p>
<p>First we add the transcript name and type information to the DET
table</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">transcript_joined</span> <span class="op">&lt;-</span> <span class="va">tx_to_gene</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">transcript_id</span>, <span class="va">transcript_name</span>, <span class="va">transcript_type</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">PE1_DETs</span> <span class="op">&lt;-</span> <span class="va">PE1_DETs</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">transcript_joined</span>, by <span class="op">=</span> <span class="st">"transcript_id"</span><span class="op">)</span></span></code></pre></div>
<p>For the gene expression level we have to convert the DEG table and do
some cutting to get the genes which present as DE and exclude possible
noise. Here we used the cutoffs of absolute log2FC higher than one and
pvalue lower than 0.05</p>
<p>So we first filter the DEG table for the significant ones and the add
it as a column on our main DET table using the isDEGsig function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PE1_DEGs_new_names_sig</span> <span class="op">&lt;-</span> <span class="va">PE1_DEGs</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pvalue</span> <span class="op">&lt;=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html" class="external-link">drop_na</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">DEGs_sig_joined</span> <span class="op">&lt;-</span> <span class="va">PE1_DEGs_new_names_sig</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">tx_to_gene</span>, by <span class="op">=</span> <span class="st">"gene_name"</span><span class="op">)</span></span>
<span><span class="va">transcript_gene_join</span> <span class="op">&lt;-</span> <span class="va">tx_to_gene</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">transcript_name</span>, <span class="va">gene_name</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">PE1_DETs_final</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/is_deg_sig.html">is_deg_sig</a></span><span class="op">(</span><span class="va">DEGs_sig_joined</span><span class="op">$</span><span class="va">transcript_name</span>, <span class="va">PE1_DETs</span><span class="op">)</span></span>
<span><span class="va">PE1_DETs_final</span> <span class="op">&lt;-</span> <span class="va">PE1_DETs_final</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">transcript_gene_join</span>, by <span class="op">=</span> <span class="st">"transcript_name"</span><span class="op">)</span></span></code></pre></div>
<p>And now we have all the tables we will need for all graphs and
analyses.</p>
<p>One detail is that the DET final table now allows us to see genes
whose transcripts are differentially expressed but their genes are not
with a simple dplyr filter.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DETs_not_DEGs</span> <span class="op">&lt;-</span> <span class="va">PE1_DETs_final</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">log2FC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span>, <span class="va">DEG_sig</span> <span class="op">==</span> <span class="st">"NO"</span><span class="op">)</span></span></code></pre></div>
<p>This table will represent cases which could be characterized as
isoform switches, when two transcripts of the same gene are expressed in
opposite directions what makes the total expression of that gene not be
significant either up or down-regulated.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="colors">Colors<a class="anchor" aria-label="anchor" href="#colors"></a>
</h2>
<p>Before we start plotting we will define a general set of colors to be
used through the entire pipeline associated with a certain type of
transcript. Here we colored all the most abundant types separately and
the less abundant on the same grey tone and name that vector
accordingly</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># TODO: @luciorq Check implementation of function `tx_type_palette()`</span></span>
<span><span class="va">fixed_tx_biotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"gene"</span>, <span class="st">"protein_coding"</span>,</span>
<span>  <span class="st">"retained_intron"</span>, <span class="st">"protein_coding_CDS_not_defined"</span>,</span>
<span>  <span class="st">"nonsense_mediated_decay"</span>, <span class="st">"lncRNA"</span>,</span>
<span>  <span class="st">"processed_pseudogene"</span>, <span class="st">"transcribed_unprocessed_pseudogene"</span>,</span>
<span>  <span class="st">"unprocessed_pseudogene"</span>, <span class="st">"non_stop_decay"</span>,</span>
<span>  <span class="st">"transcribed_unitary_pseudogene"</span>, <span class="st">"pseudogene"</span>,</span>
<span>  <span class="st">"unitary_pseudogene"</span>, <span class="st">"processed_transcript"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">tx_type_color_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"#fb8072"</span>, <span class="st">"#a6d854"</span>,</span>
<span>  <span class="st">"#8da0cb"</span>, <span class="st">"#fc8d62"</span>,</span>
<span>  <span class="st">"#66c2a5"</span>, <span class="st">"#e78ac3"</span>,</span>
<span>  <span class="st">"#ffd92f"</span>, <span class="st">"#e5c494"</span>,</span>
<span>  <span class="st">"#d9d9d9"</span>, <span class="st">"#d9d9d9"</span>,</span>
<span>  <span class="st">"#d9d9d9"</span>, <span class="st">"#ffffb3"</span>,</span>
<span>  <span class="st">"#d9d9d9"</span>, <span class="st">"#d9d9d9"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tx_type_color_names</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">fixed_tx_biotypes</span></span>
<span></span>
<span><span class="va">tx_type_color_names</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="log2fc-plot">Log2FC Plot<a class="anchor" aria-label="anchor" href="#log2fc-plot"></a>
</h2>
<p>The simplest and first plot on this tutorial will be a log2FC plot,
this plot will compare the foldchange of case vs control from the gene,
to that of its transcripts adding to that the significance
information.</p>
<p>For that we will make a combined version of the DEG table with the
DET table using the function join_DEG_DET.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DEG_DET_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/join_DEG_DET.html">join_DEG_DET</a></span><span class="op">(</span><span class="va">PE1_DEGs</span>, <span class="va">PE1_DETs_final</span>, <span class="fl">1</span>, <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DEG_DET_table</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">##   id             log2FC pvalue qvalue name  transcript_type gene_name abs_log2FC</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> ENSG00000000… -<span style="color: #BB0000;">0.320</span>  0.048<span style="text-decoration: underline;">5</span>  0.125 TSPA… gene            TSPAN6        0.320 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ENSG00000000…  0.381  0.107   0.222 TNMD  gene            TNMD          0.381 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> ENSG00000000… -<span style="color: #BB0000;">0.069</span><span style="color: #BB0000; text-decoration: underline;">3</span> 0.260   0.417 DPM1  gene            DPM1          0.069<span style="text-decoration: underline;">3</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> ENSG00000000…  0.107  0.386   0.547 SCYL3 gene            SCYL3         0.107 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ENSG00000000…  0.162  0.229   0.380 C1or… gene            C1orf112      0.162 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> ENSG00000000… -<span style="color: #BB0000;">0.162</span>  0.160   0.296 FGR   gene            FGR           0.162 </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: significance &lt;chr&gt;, DEG_sig &lt;chr&gt;</span></span></span></code></pre>
<p>Now you just use the <code>plotLog2FC</code> for any gene you would
like. The function also works well with a small vector of
<code>gene_names</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># selected_gene = "RBPJ"</span></span>
<span><span class="fu"><a href="../reference/plot_log2FC.html">plot_log2FC</a></span><span class="op">(</span><span class="va">DEG_DET_table</span>, <span class="st">"RBPJ"</span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Work here to look better? or just remove for now</span></span>
<span><span class="co"># the best would be a loop that goes over each one of a list and</span></span>
<span><span class="co"># plots them in a folder the default could be the DET not deg table</span></span>
<span><span class="va">plot_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_log2FC.html">plot_log2FC</a></span><span class="op">(</span><span class="va">DEG_DET_table</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RBPJ"</span>, <span class="st">"EGFR"</span>, <span class="st">"PNCK"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_obj</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">45</span>, vjust <span class="op">=</span> <span class="fl">1</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="profile-plot">Profile Plot<a class="anchor" aria-label="anchor" href="#profile-plot"></a>
</h2>
<p>Another good more quantifiable way to visualize this switch is using
the Transcript per Million Count of each transcript, compared to those
of the gene between the case and control conditions. For this we use a
profile plot that in one size plots the values of TPM for the case
conditions and in the other the value of the TPM for the control
conditions.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DEG_DET_table</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">transcript_type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"gene"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 70,870 × 10</span></span></span>
<span><span class="co">##    id          log2FC  pvalue  qvalue name  transcript_type gene_name abs_log2FC</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> ENST000006…  0.640 3.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 5.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> ISG1… protein_coding  ISG15          0.640</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> ENST000003…  0.534 1.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 7.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> AGRN… protein_coding  AGRN           0.534</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ENST000004… -<span style="color: #BB0000;">0.530</span> 5.81<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 1.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> AGRN… retained_intron AGRN           0.530</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> ENST000004…  1.83  1.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 6.81<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> RNF2… protein_coding  RNF223         1.83 </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> ENST000004…  0.415 2.20<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 4.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> TNFR… retained_intron TNFRSF4        0.415</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> ENST000003…  0.374 2.27<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 4.48<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> TNFR… protein_coding  TNFRSF4        0.374</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> ENST000003…  0.479 6.70<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 2.09<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> TAS1… protein_coding  TAS1R3         0.479</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> ENST000003…  0.217 3.82<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 6.06<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> DVL1… protein_coding  DVL1           0.217</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> ENST000003…  0.570 8.12<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 2.37<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> DVL1… protein_coding  DVL1           0.570</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ENST000006…  1.91  4.34<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 6.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> DVL1… protein_coding  DVL1           1.91 </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 70,860 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 2 more variables: significance &lt;chr&gt;, DEG_sig &lt;chr&gt;</span></span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">DEG_DET_table</span><span class="op">$</span><span class="va">transcript_type</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "gene"                               "protein_coding"                    </span></span>
<span><span class="co">##  [3] "retained_intron"                    "processed_transcript"              </span></span>
<span><span class="co">##  [5] "lncRNA"                             "nonsense_mediated_decay"           </span></span>
<span><span class="co">##  [7] "processed_pseudogene"               "unprocessed_pseudogene"            </span></span>
<span><span class="co">##  [9] "transcribed_unprocessed_pseudogene" "non_stop_decay"                    </span></span>
<span><span class="co">## [11] "transcribed_unitary_pseudogene"     "pseudogene"                        </span></span>
<span><span class="co">## [13] "unitary_pseudogene"</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">profile_data_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prepare_profile_data.html">prepare_profile_data</a></span><span class="op">(</span></span>
<span>  txi_transcript <span class="op">=</span> <span class="va">PE1_counts</span>,</span>
<span>  tx_to_gene <span class="op">=</span> <span class="va">tx_to_gene</span>,</span>
<span>  sample_metadata <span class="op">=</span> <span class="va">sample_table</span>,</span>
<span>  de_result_gene <span class="op">=</span> <span class="va">PE1_DEGs</span>,</span>
<span>  de_result_transcript <span class="op">=</span> <span class="va">PE1_DETs</span>,</span>
<span>  var <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>  var_levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"treatment"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># # Create output directory if don't exist yet</span></span>
<span><span class="co"># path_to_save &lt;- "results/profile_plot"</span></span>
<span><span class="co"># fs::dir_create(path_to_save)</span></span>
<span></span>
<span><span class="va">profile_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_tx_expr.html">plot_tx_expr</a></span><span class="op">(</span></span>
<span>  genes_to_plot <span class="op">=</span> <span class="st">"RBPJ"</span>,</span>
<span>  profile_data <span class="op">=</span> <span class="va">profile_data_df</span></span>
<span><span class="op">)</span></span>
<span><span class="va">profile_plot</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">profile_data_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genename</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"IL2RA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 7</span></span></span>
<span><span class="co">##   genename condition mean_TPM    SD transcript_type DE    parent_gene</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> IL2RA    control       32.5 14.5  gene            Yes   genename   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> IL2RA    treatment     11.8  9.04 gene            Yes   genename</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">profile_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_tx_expr.html">plot_tx_expr</a></span><span class="op">(</span></span>
<span>  genes_to_plot <span class="op">=</span> <span class="st">"IL2RA"</span>,</span>
<span>  profile_data <span class="op">=</span> <span class="va">profile_data_df</span></span>
<span><span class="op">)</span></span>
<span><span class="va">profile_plot</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="functional-transcript-enrichment">Functional Transcript Enrichment<a class="anchor" aria-label="anchor" href="#functional-transcript-enrichment"></a>
</h2>
<p>One of the biggest caveats for transcript level analysis is that in
many times is hard to extract biologically relevant information from so
much data. Instead of having a final table with 900 genes you get a
table with over 3.000 transcripts after the differential expression cut.
The next step for gene-level DE would be functional enrichment or
assigning the genes to metabolic pathways those may be regulating.
Unfortunately are no comprehensive datasets for pathways transcripts may
be regulating and the gene level analyses normally loses the difference
between those transcripts which can produce proteins (protein_coding)
from canonical translation pathways and those which cannot. To solve
this problem we developed a method of expanding the known .gmts for
transcript information and then separately enrich each selected category
of transcript. Between the alternative spliced isoforms of that do not
code for canonical proteins the most abundant are those classified as
<em>Nonsense-mediated decay</em>, that have a premature stop codon which
is subject to targeted degradation and the <em>Protein coding CDS not
defined</em> (formerly identified as “processed transcript”), which, for
any reason, do not possess a complete Open Reading Frame. Inside the
processed transcript category the one with the highest count are the
<em>Retained introns</em>, sequences which retain an intronic portion
after their processing.</p>
<p>These three categories are the most abundant in those transcripts
which arise from the alternative splicing of a protein coding gene and
these three will be the main focus for our enrichment and further
graphs.</p>
<p>So first we choose a .gmt to be used for the enrichment, in this case
we loaded a human reactome gene list from MSigDB called c2. But any gene
list in <a href="https://docs.gsea-msigdb.org/#GSEA/Data_Formats/" class="external-link">GMT
format</a> works here. The gmt is loaded on the fgsea format with lists
for each biological process.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span>, <span class="st">"c2.cp.reactome.v2023.1.Hs.symbols.gmt.txt"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## /private/var/folders/2q/937_bkg10svdwx1x00prs9nm0000gn/T/RtmpwoIG2f/temp_libpathedcf4f41b3f9/isoformic/extdata/c2.cp.reactome.v2023.1.Hs.symbols.gmt.txt</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genesets_list</span> <span class="op">&lt;-</span> <span class="fu">fgsea</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/fgsea/man/gmtPathways.html" class="external-link">gmtPathways</a></span><span class="op">(</span></span>
<span>  gmt.file <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"isoformic"</span>, <span class="st">"extdata"</span>, <span class="st">"c2.cp.reactome.v2023.1.Hs.symbols.gmt.txt"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">genesets_list</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## List of 5</span></span>
<span><span class="co">##  $ REACTOME_INTERLEUKIN_6_SIGNALING        : chr [1:11] "JAK2" "TYK2" "CBL" "STAT1" ...</span></span>
<span><span class="co">##  $ REACTOME_APOPTOSIS                      : chr [1:179] "BAD" "CFLAR" "PSMB1" "PSMC4" ...</span></span>
<span><span class="co">##  $ REACTOME_HEMOSTASIS                     : chr [1:678] "FGR" "CD99" "TFPI" "KDM1A" ...</span></span>
<span><span class="co">##  $ REACTOME_INTRINSIC_PATHWAY_FOR_APOPTOSIS: chr [1:55] "BAD" "BID" "BAK1" "CASP8" ...</span></span>
<span><span class="co">##  $ REACTOME_MAPK3_ERK1_ACTIVATION          : chr [1:10] "JAK2" "MAPK3" "TYK2" "IL6ST" ...</span></span></code></pre>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genesets_list</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "JAK2"  "TYK2"  "CBL"   "STAT1" "IL6ST" "IL6"</span></span></code></pre>
<p>Visualize how is our table before running</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">PE1_DETs_final</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 8</span></span></span>
<span><span class="co">##   transcript_id   log2FC  pvalue  qvalue transcript_name transcript_type DEG_sig</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> ENST0000064952…  0.640 3.11<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 5.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> ISG15-203       protein_coding  YES    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> ENST0000037937…  0.534 1.29<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 7.07<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> AGRN-201        protein_coding  YES    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> ENST0000046111… -<span style="color: #BB0000;">0.530</span> 5.81<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> 1.91<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> AGRN-202        retained_intron YES    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> ENST0000045346…  1.83  1.66<span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> 6.81<span style="color: #949494;">e</span><span style="color: #BB0000;">-4</span> RNF223-201      protein_coding  YES    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> ENST0000049786…  0.415 2.20<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 4.40<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> TNFRSF4-203     retained_intron YES    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> ENST0000037923…  0.374 2.27<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> 4.48<span style="color: #949494;">e</span><span style="color: #BB0000;">-1</span> TNFRSF4-201     protein_coding  YES    </span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 1 more variable: gene_name &lt;chr&gt;</span></span></span></code></pre>
<p>Then you run the <code><a href="../reference/run_enrichment.html">run_enrichment()</a></code> function it needs your
DETs final table, the gene set list and a p-value cutoff to be used. It
will generate a table of enrichment but with an extra column
“experiment”.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_enrichment.html">run_enrichment</a></span><span class="op">(</span></span>
<span>  det_df <span class="op">=</span> <span class="va">PE1_DETs_final</span>,</span>
<span>  genesets_list <span class="op">=</span> <span class="va">genesets_list</span>,</span>
<span>  tx_to_gene <span class="op">=</span> <span class="va">tx_to_gene</span>,</span>
<span>  pval_cutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enrichment_df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                         pathway         pval</span></span>
<span><span class="co">##                                                          &lt;char&gt;        &lt;num&gt;</span></span>
<span><span class="co">## 1:              REACTOME_ABC_FAMILY_PROTEINS_MEDIATED_TRANSPORT 0.0025381836</span></span>
<span><span class="co">## 2:                           REACTOME_ABC_TRANSPORTER_DISORDERS 0.0182832492</span></span>
<span><span class="co">## 3: REACTOME_ACROSOME_REACTION_AND_SPERM_OOCYTE_MEMBRANE_BINDING 0.0224136073</span></span>
<span><span class="co">## 4:       REACTOME_ACTIVATED_NTRK2_SIGNALS_THROUGH_FRS2_AND_FRS3 0.0437500000</span></span>
<span><span class="co">## 5:                REACTOME_ACTIVATED_NTRK2_SIGNALS_THROUGH_PI3K 0.0006616736</span></span>
<span><span class="co">## 6:                 REACTOME_ACTIVATED_NTRK2_SIGNALS_THROUGH_RAS 0.0041758237</span></span>
<span><span class="co">##          padj   log2err        ES      NES  size</span></span>
<span><span class="co">##         &lt;num&gt;     &lt;num&gt;     &lt;num&gt;    &lt;num&gt; &lt;int&gt;</span></span>
<span><span class="co">## 1: 0.04818654 0.4317077 0.3725212 1.465466   263</span></span>
<span><span class="co">## 2: 0.14031331 0.3524879 0.3614713 1.381439   196</span></span>
<span><span class="co">## 3: 0.16008590 0.3524879 0.8512491 1.565424     5</span></span>
<span><span class="co">## 4: 0.24806701 0.2712886 0.5351451 1.488537    26</span></span>
<span><span class="co">## 5: 0.02426137 0.4772708 0.7292574 1.880501    19</span></span>
<span><span class="co">## 6: 0.06500103 0.4070179 0.6460168 1.777428    24</span></span>
<span><span class="co">##                                                        leadingEdge</span></span>
<span><span class="co">##                                                             &lt;list&gt;</span></span>
<span><span class="co">## 1:     OS9-209,PSMD7-204,PSMA5-206,UBC-205,ABCB1-201,UBA52-201,...</span></span>
<span><span class="co">## 2:    OS9-209,PSMD7-204,PSMA5-206,UBC-205,UBA52-201,ERLIN2-206,...</span></span>
<span><span class="co">## 3:                                 CD9-208,CD9-216,CD9-212,CD9-204</span></span>
<span><span class="co">## 4:     NTRK2-208,NTRK2-204,NTF4-203,NTF4-201,FRS2-206,FRS3-201,...</span></span>
<span><span class="co">## 5: NTRK2-208,PIK3R1-202,PIK3R1-216,NTRK2-204,NTF4-203,GAB1-201,...</span></span>
<span><span class="co">## 6:                  NTRK2-208,NTRK2-204,NTF4-203,NTF4-201,SHC1-208</span></span>
<span><span class="co">##        experiment</span></span>
<span><span class="co">##            &lt;char&gt;</span></span>
<span><span class="co">## 1: protein_coding</span></span>
<span><span class="co">## 2: protein_coding</span></span>
<span><span class="co">## 3: protein_coding</span></span>
<span><span class="co">## 4: protein_coding</span></span>
<span><span class="co">## 5: protein_coding</span></span>
<span><span class="co">## 6: protein_coding</span></span></code></pre>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">genesets_list</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "REACTOME_INTERLEUKIN_6_SIGNALING"                                                           </span></span>
<span><span class="co">## [2] "REACTOME_APOPTOSIS"                                                                         </span></span>
<span><span class="co">## [3] "REACTOME_HEMOSTASIS"                                                                        </span></span>
<span><span class="co">## [4] "REACTOME_INTRINSIC_PATHWAY_FOR_APOPTOSIS"                                                   </span></span>
<span><span class="co">## [5] "REACTOME_MAPK3_ERK1_ACTIVATION"                                                             </span></span>
<span><span class="co">## [6] "REACTOME_TRANSLESION_SYNTHESIS_BY_Y_FAMILY_DNA_POLYMERASES_BYPASSES_LESIONS_ON_DNA_TEMPLATE"</span></span></code></pre>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">enrichment_df</span><span class="op">$</span><span class="va">experiment</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "protein_coding"          "unproductive"           </span></span>
<span><span class="co">## [3] "retained_intron"         "processed_transcript"   </span></span>
<span><span class="co">## [5] "nonsense_mediated_decay"</span></span></code></pre>
<p>This experiment column has five possible values: Protein-coding:
which is the enrichment associated with the transcripts categorized as
protein coding. Unproductive: This is a term that will be used moving
forward to combine those three categories of alternative spliced
isoforms transcribed by coding genes. The authors are aware that
biologically this term is deprecated since those kind of transcripts can
produce peptides from alternative translation pathways. So here
unproductive should be read as virtually incapable of producing the
protein that is associated with that gene. As interpretation, we made
this category to find pathways which are not being regulated on our
coding data, but by the unproductive transcripts.</p>
<p>We also added three categories which are the individual alternative
spliced types and the pathways regulated by those for specific analysis.
In a very deep transcriptome the individual enrichment from those
categories can also lead to promising insights.</p>
<p>Plotting the enrichment</p>
<p>We used a LollipopPlot to plot all the enrichments side by side with
the size of each pathway as the radius of the circles and the
transparency is if that pathway passes on the desired cutoff. First we
plot for only Protein_coding versus Unproductive with a very extringent
NES cutoff.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>    <span class="va">experiment</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"protein_coding"</span>, <span class="st">"unproductive"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">NES</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">30</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pathway</span>, <span class="va">experiment</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">NES</span>, size <span class="op">=</span> <span class="va">size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"mako"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-31-1.png" width="1440"></p>
<p>And now the specific unproductive subtypes</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enrichment_df</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">experiment</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"protein_coding"</span>, <span class="st">"unproductive"</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">NES</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1.5</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">padj</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">20</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">pathway</span>, <span class="va">experiment</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">NES</span>, size <span class="op">=</span> <span class="va">size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_minimal</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridis/reference/scale_viridis.html" class="external-link">scale_color_viridis</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"magma"</span>, direction <span class="op">=</span> <span class="op">-</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-32-1.png" width="1440"></p>
</div>
<div class="section level2">
<h2 id="genomic-context-plot">Genomic Context Plot<a class="anchor" aria-label="anchor" href="#genomic-context-plot"></a>
</h2>
<p>One of the main issues we arrived at the start of the isoform level
analysis, was that there was no easy direct way to visualize
transcript-types if compared one to another, and using the
transcript-type and the transcript per million information.</p>
<p>Most of the alignment plots today use the outputs from .bam/.sam
files that align directly to the genome making us lose the
transcript-type information and increasing considerably the processing
time for any analysis for the size of the files and the time it takes to
re-align.</p>
<p>To solve this problem we used a more direct approach which allows us
to visualize the difference of introns and exons between the transcript,
the types of those transcripts and how much they were counted according
to the pseudo-alignment; but in turn it loses the read alignment count
proportion. This alignment count unfortunately requires running
alignment softwares and dealing with .sam and .bam files which will not
be covered on this tutorial. We called this plot the genomic context
plot and it takes inspiration from the way Ensembl shows it transcripts
on their genome browser.</p>
<p>This specific plot requires a GFF file that can also be downloaded
from GENCODE to be included in the file path on next function. This GFF
file needs to be downloaded on the accurate version for your
transcriptome, in this case v34.</p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">context_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_context_data.html">create_context_data</a></span><span class="op">(</span></span>
<span>  gff_file <span class="op">=</span> <span class="st">"data-raw/gencode.v34.annotation.gff3.gz"</span>,</span>
<span>  organism <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>  orgdb_package <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span>,</span>
<span>  bsgenome_package <span class="op">=</span> <span class="st">"BSgenome.Hsapiens.NCBI.GRCh38"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Plotting the context of the FLT1 isoforms.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_context.html">plot_genomic_context</a></span><span class="op">(</span></span>
<span>  gene_name <span class="op">=</span> <span class="st">"FLT1"</span>,</span>
<span>  context_data <span class="op">=</span> <span class="va">context_data</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-34-1.png" width="700"></p>
<div class="section level4">
<h4 id="protein-coding-example">Protein coding example<a class="anchor" aria-label="anchor" href="#protein-coding-example"></a>
</h4>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_context.html">plot_genomic_context</a></span><span class="op">(</span></span>
<span>  gene_name <span class="op">=</span> <span class="st">"EGFR"</span>,</span>
<span>  context_data <span class="op">=</span> <span class="va">context_data</span>,</span>
<span>  y_offset <span class="op">=</span> <span class="fl">4.5</span>,</span>
<span>  height_offset <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<p>Expanding the context around the genomic region of interest.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_context.html">plot_genomic_context</a></span><span class="op">(</span></span>
<span>  gene_name <span class="op">=</span> <span class="st">"XIST"</span>,</span>
<span>  context_data <span class="op">=</span> <span class="va">context_data</span>,</span>
<span>  y_offset <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  height_offset <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  upstream_offset <span class="op">=</span> <span class="fl">100000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_genomic_context.html">plot_genomic_context</a></span><span class="op">(</span></span>
<span>  gene_name <span class="op">=</span> <span class="st">"TSIX"</span>,</span>
<span>  context_data <span class="op">=</span> <span class="va">context_data</span>,</span>
<span>  y_offset <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  height_offset <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  downstream_offset <span class="op">=</span> <span class="fl">100000</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="isoformic_intro_files/figure-html/unnamed-chunk-37-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sequoia 15.6</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/New_York</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] GenomicRanges_1.61.1 Seqinfo_0.99.2       IRanges_2.43.0      </span></span>
<span><span class="co">##  [4] S4Vectors_0.47.0     BiocGenerics_0.55.1  generics_0.1.4      </span></span>
<span><span class="co">##  [7] ggplot2_4.0.0        stringr_1.5.2        tidyr_1.3.1         </span></span>
<span><span class="co">## [10] dplyr_1.1.4          readr_2.1.5          fs_1.6.6            </span></span>
<span><span class="co">## [13] isoformic_0.1.2     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] strawr_0.0.92                           </span></span>
<span><span class="co">##   [2] RColorBrewer_1.1-3                      </span></span>
<span><span class="co">##   [3] jsonlite_2.0.0                          </span></span>
<span><span class="co">##   [4] magrittr_2.0.4                          </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.61.6                  </span></span>
<span><span class="co">##   [6] farver_2.1.2                            </span></span>
<span><span class="co">##   [7] rmarkdown_2.30                          </span></span>
<span><span class="co">##   [8] BiocIO_1.19.0                           </span></span>
<span><span class="co">##   [9] ragg_1.5.0                              </span></span>
<span><span class="co">##  [10] vctrs_0.6.5                             </span></span>
<span><span class="co">##  [11] memoise_2.0.1                           </span></span>
<span><span class="co">##  [12] Rsamtools_2.25.2                        </span></span>
<span><span class="co">##  [13] RCurl_1.98-1.17                         </span></span>
<span><span class="co">##  [14] rstatix_0.7.2                           </span></span>
<span><span class="co">##  [15] htmltools_0.5.8.1                       </span></span>
<span><span class="co">##  [16] S4Arrays_1.9.1                          </span></span>
<span><span class="co">##  [17] forcats_1.0.1                           </span></span>
<span><span class="co">##  [18] progress_1.2.3                          </span></span>
<span><span class="co">##  [19] AnnotationHub_3.99.6                    </span></span>
<span><span class="co">##  [20] curl_7.0.0                              </span></span>
<span><span class="co">##  [21] broom_1.0.10                            </span></span>
<span><span class="co">##  [22] Rhdf5lib_1.31.0                         </span></span>
<span><span class="co">##  [23] rhdf5_2.53.3                            </span></span>
<span><span class="co">##  [24] gridGraphics_0.5-1                      </span></span>
<span><span class="co">##  [25] SparseArray_1.9.1                       </span></span>
<span><span class="co">##  [26] Formula_1.2-5                           </span></span>
<span><span class="co">##  [27] sass_0.4.10                             </span></span>
<span><span class="co">##  [28] bslib_0.9.0                             </span></span>
<span><span class="co">##  [29] htmlwidgets_1.6.4                       </span></span>
<span><span class="co">##  [30] desc_1.4.3                              </span></span>
<span><span class="co">##  [31] httr2_1.2.1                             </span></span>
<span><span class="co">##  [32] cachem_1.1.0                            </span></span>
<span><span class="co">##  [33] GenomicAlignments_1.45.2                </span></span>
<span><span class="co">##  [34] lifecycle_1.0.4                         </span></span>
<span><span class="co">##  [35] pkgconfig_2.0.3                         </span></span>
<span><span class="co">##  [36] Matrix_1.7-4                            </span></span>
<span><span class="co">##  [37] R6_2.6.1                                </span></span>
<span><span class="co">##  [38] fastmap_1.2.0                           </span></span>
<span><span class="co">##  [39] GenomeInfoDbData_1.2.14                 </span></span>
<span><span class="co">##  [40] MatrixGenerics_1.21.0                   </span></span>
<span><span class="co">##  [41] digest_0.6.37                           </span></span>
<span><span class="co">##  [42] AnnotationDbi_1.71.1                    </span></span>
<span><span class="co">##  [43] textshaping_1.0.3                       </span></span>
<span><span class="co">##  [44] RSQLite_2.4.3                           </span></span>
<span><span class="co">##  [45] org.Hs.eg.db_3.21.0                     </span></span>
<span><span class="co">##  [46] ggpubr_0.6.1                            </span></span>
<span><span class="co">##  [47] filelock_1.0.3                          </span></span>
<span><span class="co">##  [48] labeling_0.4.3                          </span></span>
<span><span class="co">##  [49] httr_1.4.7                              </span></span>
<span><span class="co">##  [50] abind_1.4-8                             </span></span>
<span><span class="co">##  [51] compiler_4.5.1                          </span></span>
<span><span class="co">##  [52] bit64_4.6.0-1                           </span></span>
<span><span class="co">##  [53] withr_3.0.2                             </span></span>
<span><span class="co">##  [54] S7_0.2.0                                </span></span>
<span><span class="co">##  [55] backports_1.5.0                         </span></span>
<span><span class="co">##  [56] BiocParallel_1.43.4                     </span></span>
<span><span class="co">##  [57] carData_3.0-5                           </span></span>
<span><span class="co">##  [58] viridis_0.6.5                           </span></span>
<span><span class="co">##  [59] DBI_1.2.3                               </span></span>
<span><span class="co">##  [60] ggsignif_0.6.4                          </span></span>
<span><span class="co">##  [61] biomaRt_2.64.0                          </span></span>
<span><span class="co">##  [62] rappdirs_0.3.3                          </span></span>
<span><span class="co">##  [63] DelayedArray_0.34.1                     </span></span>
<span><span class="co">##  [64] rjson_0.2.23                            </span></span>
<span><span class="co">##  [65] tools_4.5.1                             </span></span>
<span><span class="co">##  [66] glue_1.8.0                              </span></span>
<span><span class="co">##  [67] restfulr_0.0.16                         </span></span>
<span><span class="co">##  [68] rhdf5filters_1.21.0                     </span></span>
<span><span class="co">##  [69] grid_4.5.1                              </span></span>
<span><span class="co">##  [70] fgsea_1.35.6                            </span></span>
<span><span class="co">##  [71] gtable_0.3.6                            </span></span>
<span><span class="co">##  [72] tzdb_0.5.0                              </span></span>
<span><span class="co">##  [73] data.table_1.17.8                       </span></span>
<span><span class="co">##  [74] hms_1.1.3                               </span></span>
<span><span class="co">##  [75] xml2_1.4.0                              </span></span>
<span><span class="co">##  [76] car_3.1-3                               </span></span>
<span><span class="co">##  [77] utf8_1.2.6                              </span></span>
<span><span class="co">##  [78] XVector_0.49.0                          </span></span>
<span><span class="co">##  [79] BiocVersion_3.22.0                      </span></span>
<span><span class="co">##  [80] ggrepel_0.9.6                           </span></span>
<span><span class="co">##  [81] pillar_1.11.1                           </span></span>
<span><span class="co">##  [82] yulab.utils_0.2.1                       </span></span>
<span><span class="co">##  [83] vroom_1.6.6                             </span></span>
<span><span class="co">##  [84] BiocFileCache_2.99.5                    </span></span>
<span><span class="co">##  [85] lattice_0.22-7                          </span></span>
<span><span class="co">##  [86] rtracklayer_1.69.1                      </span></span>
<span><span class="co">##  [87] bit_4.6.0                               </span></span>
<span><span class="co">##  [88] tidyselect_1.2.1                        </span></span>
<span><span class="co">##  [89] Biostrings_2.77.2                       </span></span>
<span><span class="co">##  [90] knitr_1.50                              </span></span>
<span><span class="co">##  [91] gridExtra_2.3                           </span></span>
<span><span class="co">##  [92] SummarizedExperiment_1.39.1             </span></span>
<span><span class="co">##  [93] xfun_0.53                               </span></span>
<span><span class="co">##  [94] Biobase_2.69.0                          </span></span>
<span><span class="co">##  [95] matrixStats_1.5.0                       </span></span>
<span><span class="co">##  [96] stringi_1.8.7                           </span></span>
<span><span class="co">##  [97] UCSC.utils_1.5.0                        </span></span>
<span><span class="co">##  [98] yaml_2.3.10                             </span></span>
<span><span class="co">##  [99] TxDb.Hsapiens.UCSC.hg38.knownGene_3.21.0</span></span>
<span><span class="co">## [100] evaluate_1.0.5                          </span></span>
<span><span class="co">## [101] codetools_0.2-20                        </span></span>
<span><span class="co">## [102] tibble_3.3.0                            </span></span>
<span><span class="co">## [103] BiocManager_1.30.26                     </span></span>
<span><span class="co">## [104] ggplotify_0.1.3                         </span></span>
<span><span class="co">## [105] cli_3.6.5                               </span></span>
<span><span class="co">## [106] systemfonts_1.3.1                       </span></span>
<span><span class="co">## [107] jquerylib_0.1.4                         </span></span>
<span><span class="co">## [108] Rcpp_1.1.0                              </span></span>
<span><span class="co">## [109] GenomeInfoDb_1.45.9                     </span></span>
<span><span class="co">## [110] dbplyr_2.5.1                            </span></span>
<span><span class="co">## [111] png_0.1-8                               </span></span>
<span><span class="co">## [112] XML_3.99-0.19                           </span></span>
<span><span class="co">## [113] parallel_4.5.1                          </span></span>
<span><span class="co">## [114] pkgdown_2.1.3                           </span></span>
<span><span class="co">## [115] blob_1.2.4                              </span></span>
<span><span class="co">## [116] prettyunits_1.2.0                       </span></span>
<span><span class="co">## [117] plyranges_1.28.0                        </span></span>
<span><span class="co">## [118] bitops_1.0-9                            </span></span>
<span><span class="co">## [119] txdbmaker_1.4.2                         </span></span>
<span><span class="co">## [120] viridisLite_0.4.2                       </span></span>
<span><span class="co">## [121] scales_1.4.0                            </span></span>
<span><span class="co">## [122] plotgardener_1.14.0                     </span></span>
<span><span class="co">## [123] purrr_1.1.0                             </span></span>
<span><span class="co">## [124] crayon_1.5.3                            </span></span>
<span><span class="co">## [125] rlang_1.1.6                             </span></span>
<span><span class="co">## [126] cowplot_1.2.0                           </span></span>
<span><span class="co">## [127] fastmatch_1.1-6                         </span></span>
<span><span class="co">## [128] KEGGREST_1.49.1</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lucio Rezende Queiroz, Lucio Rezende Queiroz, Izabela Mamede Conceicao, Luigi Marchionni, Gloria Franco.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
