<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Preprocessing Short-Read Sequencing Files for Isoformic • isoformic</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="isoformic-preprocess_files/libs/quarto-html/popper.min.js"></script><script src="isoformic-preprocess_files/libs/quarto-html/tippy.umd.min.js"></script><link href="isoformic-preprocess_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="isoformic-preprocess_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Preprocessing Short-Read Sequencing Files for Isoformic">
<meta property="og:image" content="https://luciorq.github.io/isoformic/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">isoformic</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2.9017</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/isoformic-preprocess.html">Preprocessing Short-Read Sequencing Files for Isoformic</a></li>
    <li><a class="dropdown-item" href="../articles/isoformic-table-based.html">Isoformic: Isoform-level biological interpretation of transcriptomic data</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/luciorq/isoformic/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Preprocessing Short-Read Sequencing Files for Isoformic</h1>

      <p class="author">Lucio Rezende QueirozIzabela Mamede Conceição</p>
      <p class="date">October 2, 2025</p>

      <small class="dont-index">Source: <a href="https://github.com/luciorq/isoformic/blob/main/vignettes/isoformic-preprocess.qmd" class="external-link"><code>vignettes/isoformic-preprocess.qmd</code></a></small>
      <div class="d-none name"><code></code></div>
    </div>






    <section class="level2"><h2 id="salmon-command-command-line-interface-anatomy">Salmon: command command line interface anatomy<a class="anchor" aria-label="anchor" href="#salmon-command-command-line-interface-anatomy"></a>
</h2>
<section class="level3"><h3 id="from-the-terminal---salmon-command-line-interface">From the Terminal - Salmon Command Line Interface<a class="anchor" aria-label="anchor" href="#from-the-terminal---salmon-command-line-interface"></a>
</h3>
<p>Below is a single-sample invocation (to be run in a Bash shell) showing the typical arguments we used when making the mini dataset used in this vignette.</p>
<p>Replace the paths with yours if you want to reproduce.</p>
<p>More thorough information can be found on: <a href="https://combine-lab.github.io/salmon/" class="external-link uri">https://combine-lab.github.io/salmon/</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Not needed as of Salmon 1.10.3:</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># --validateMappings \               # selective-alignment (more specific, recommended)</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">salmon</span> quant <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--libType</span> A <span class="dt">\ </span>                     <span class="co"># A -&gt; autodetect library type from the reads</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--index</span> /path/to/index <span class="dt">\ </span>          <span class="co"># Transcriptome index (here: index built from the FASTA annotation file)</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--mates1</span> /path/SRR..._1.fastq <span class="dt">\ </span>   <span class="co"># R1 FASTQ</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--mates2</span> /path/SRR..._2.fastq <span class="dt">\ </span>   <span class="co"># R2 FASTQ</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--output</span> /path/SRR..._quant <span class="dt">\ </span>     <span class="co"># Output folder (contains quant.sf, aux_info/)</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--threads</span> 8 <span class="dt">\ </span>                     <span class="co"># Number of parallel threads used</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--disableChainingHeuristic</span> <span class="dt">\ </span>      <span class="co"># improve mapping sensitivity, but slower</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--allowDovetail</span> <span class="dt">\</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">--softclipOverhangs</span> <span class="dt">\ </span>             <span class="co"># softclip reads that overhang transcripts start and end</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--dumpEq</span> <span class="dt">\</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">--dumpEqWeights</span> <span class="dt">\ </span>                 <span class="co"># dump equivalent class information, needed by Terminus</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--posBias</span> <span class="at">--seqBias</span> <span class="at">--gcBias</span> <span class="dt">\ </span>    <span class="co"># bias corrections (positional, sequence, GC)</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--numGibbsSamples</span> 100              <span class="co"># draw inferential replicates; required for fishpond/Swish</span></span></code></pre></div>
</section><section class="level3"><h3 id="from-r---condathis-wrapper-for-salmon">From R - <code>condathis</code> wrapper for Salmon<a class="anchor" aria-label="anchor" href="#from-r---condathis-wrapper-for-salmon"></a>
</h3>
<p>Caveats: currently <a href="https://anaconda.org/bioconda/salmon" class="external-link">Salmon</a> is only available for macOS and Linux on the <a href="https://bioconda.github.io/" class="external-link">Bioconda</a> channel.</p>
<p>Therefore, this example will not work natively on Windows. But you can use WSL2 (Windows Subsystem for Linux) on Windows 10/11 to run it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/luciorq/isoformic" class="external-link">isoformic</a></span><span class="op">)</span></span>
<span><span class="va">tx_fasta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_reference.html">download_reference</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"49"</span>, file_type <span class="op">=</span> <span class="st">"fasta"</span><span class="op">)</span></span>
<span><span class="va">genome_fasta_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_reference.html">download_reference</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"49"</span>, file_type <span class="op">=</span> <span class="st">"genome_fasta"</span><span class="op">)</span></span>
<span><span class="va">gff_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_reference.html">download_reference</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"49"</span>, file_type <span class="op">=</span> <span class="st">"gff"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">base_dir</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="st">"data-raw"</span>, <span class="st">"vignette_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">index_path</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">base_dir</span>, <span class="st">"salmon_index_k15"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/salmon_index.html">salmon_index</a></span><span class="op">(</span></span>
<span>  fasta_path <span class="op">=</span> <span class="va">tx_fasta_path</span>,</span>
<span>  index_path <span class="op">=</span> <span class="va">index_path</span>,</span>
<span>  kmer_len <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  num_threads <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  is_gencode <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  decoy_fasta <span class="op">=</span> <span class="va">genome_fasta_path</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fastq_files</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="st">"data-raw/mini_data/sub_fastq/"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_file.html" class="external-link">path_file</a></span><span class="op">(</span><span class="va">fastq_files</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="st">"_sub_[1-2].fastq.gz$"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">sample_names</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">sample_name</span> <span class="op">&lt;-</span> <span class="va">sample_names</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span></span>
<span>  <span class="va">salmon_output</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">base_dir</span>, <span class="st">"salmon_quants"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"_quant"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">fastq_r1</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="st">"data-raw/mini_data/sub_fastq/"</span>, regexp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"_sub_1.fastq.gz$"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">fastq_r2</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="st">"data-raw/mini_data/sub_fastq/"</span>, regexp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample_name</span>, <span class="st">"_sub_2.fastq.gz$"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="co"># fs::file_exists(fastq_r1)</span></span>
<span>  <span class="co"># fs::file_exists(fastq_r2)</span></span>
<span></span>
<span>  <span class="fu"><a href="../reference/salmon_quant.html">salmon_quant</a></span><span class="op">(</span></span>
<span>    index_path <span class="op">=</span> <span class="va">index_path</span>,</span>
<span>    input_r1 <span class="op">=</span> <span class="va">fastq_r1</span>,</span>
<span>    input_r2 <span class="op">=</span> <span class="va">fastq_r2</span>,</span>
<span>    output_dir <span class="op">=</span> <span class="va">salmon_output</span>,</span>
<span>    num_threads <span class="op">=</span> <span class="fl">8</span>,</span>
<span>    num_gibbs <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    min_score_fraction <span class="op">=</span> <span class="st">"0.65"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
</section></section><section class="level2"><h2 id="load-data-and-create-isoformicexperiment-object">Load Data and Create IsoformicExperiment Object<a class="anchor" aria-label="anchor" href="#load-data-and-create-isoformicexperiment-object"></a>
</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/tximeta" class="external-link">tximeta</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://thelovelab.github.io/fishpond" class="external-link">fishpond</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/thelovelab/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/luciorq/isoformic" class="external-link">isoformic</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Where extdata lives</span></span>
<span><span class="va">extdata_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"isoformic"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">quants_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata_dir</span>, <span class="st">"mini_quants"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gff_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/download_reference.html">download_reference</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"49"</span>, file_type <span class="op">=</span> <span class="st">"gff"</span><span class="op">)</span></span>
<span><span class="co">#&gt; ✔ '/home/runner/.cache/R/isoformic/gencode.v49.chr_patch_hapl_scaff.annotation.gff3.gz'</span></span>
<span><span class="co">#&gt;   successfully downloaded.</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="load-salmon-quantification-with-tximport">Load Salmon Quantification with tximport<a class="anchor" aria-label="anchor" href="#load-salmon-quantification-with-tximport"></a>
</h2>
<p>Point to Salmon’s <code>quant.sf</code> output files.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># discover sample folders like SRR11498039_quant, etc.</span></span>
<span><span class="va">sample_dirs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.dirs</a></span><span class="op">(</span><span class="va">quants_dir</span>, full.names <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># keep only *_quant and strip the suffix for sample names</span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_quant$"</span>, <span class="st">""</span>, <span class="va">sample_dirs</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">sample_names</span><span class="op">)</span> <span class="op">==</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># define groups: "control" = first 3, "treatment" = last 3 (sorted by sample ID)</span></span>
<span><span class="va">sample_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="va">sample_names</span><span class="op">)</span></span>
<span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="fl">3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"treatment"</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">quant_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">quants_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">sample_names</span>, <span class="st">"_quant"</span><span class="op">)</span>, <span class="st">"quant.sf"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">sample_names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">quant_files</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">col_data</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span></span>
<span>  files <span class="op">=</span> <span class="va">quant_files</span>,</span>
<span>  names <span class="op">=</span> <span class="va">sample_names</span>,</span>
<span>  condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">condition</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"treatment"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">col_data</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 3</span></span>
<span><span class="co">#&gt;   files                                                          names condition</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                                                          &lt;chr&gt; &lt;fct&gt;    </span></span>
<span><span class="co">#&gt; 1 /home/runner/work/_temp/Library/isoformic/extdata/mini_quants… SRR1… control  </span></span>
<span><span class="co">#&gt; 2 /home/runner/work/_temp/Library/isoformic/extdata/mini_quants… SRR1… control  </span></span>
<span><span class="co">#&gt; 3 /home/runner/work/_temp/Library/isoformic/extdata/mini_quants… SRR1… control  </span></span>
<span><span class="co">#&gt; 4 /home/runner/work/_temp/Library/isoformic/extdata/mini_quants… SRR1… treatment</span></span>
<span><span class="co">#&gt; 5 /home/runner/work/_temp/Library/isoformic/extdata/mini_quants… SRR1… treatment</span></span>
<span><span class="co">#&gt; 6 /home/runner/work/_temp/Library/isoformic/extdata/mini_quants… SRR1… treatment</span></span></code></pre></div>
</div>
<p>Import transcriptomics data with tximport.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_tx</span> <span class="op">&lt;-</span> <span class="fu">tximeta</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tximeta/man/tximeta.html" class="external-link">tximeta</a></span><span class="op">(</span></span>
<span>  coldata <span class="op">=</span> <span class="va">col_data</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"salmon"</span>,</span>
<span>  txOut <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  skipMeta <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  useHub <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ignoreAfterBar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  skipSeqinfo <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; importing salmon quantification files</span></span>
<span><span class="co">#&gt; reading in files with read_tsv</span></span>
<span><span class="co">#&gt; 1 2 3 4 5 6</span></span>
<span></span>
<span><span class="va">mini_inf_reps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/mini_inf_reps.rds"</span>, package <span class="op">=</span> <span class="st">"isoformic"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">se_tx</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">se_tx</span><span class="op">)</span>, <span class="va">mini_inf_reps</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Convert to IsoformicExperiment object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">iso</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_isoformic.html">as_isoformic</a></span><span class="op">(</span><span class="va">se_tx</span>, <span class="va">gff_path</span><span class="op">)</span></span>
<span><span class="va">iso</span></span>
<span><span class="co">#&gt; &lt;IsoformicExperiment&gt;</span></span>
<span><span class="co">#&gt;  Experiment Name: 778d392d53b1394c97b69c3aec6ba177</span></span>
<span><span class="co">#&gt;  Data Path: /home/runner/.cache/R/isoformic/778d392d53b1394c97b69c3aec6ba177</span></span>
<span><span class="co">#&gt;  Assays: counts, abundance, length, infRep1, infRep2, infRep3, infRep4, infRep5, infRep6, infRep7, infRep8, infRep9, infRep10, infRep11, infRep12, infRep13, infRep14, infRep15, infRep16, infRep17, infRep18, infRep19, infRep20</span></span>
<span><span class="co">#&gt;  Samples: 6</span></span>
<span><span class="co">#&gt;  Transcripts in Assay: 111</span></span>
<span><span class="co">#&gt;  Annotation Name: gencode_v49_chr_patch_hapl_scaff_annotation_gff3</span></span>
<span><span class="co">#&gt;  Annotation Path: /home/runner/.cache/R/isoformic/gencode_v49_chr_patch_hapl_scaff_annotation_gff3</span></span>
<span><span class="co">#&gt;  Transcripts in Annotation: 533740</span></span>
<span><span class="co">#&gt;  Genes in Annotation: 86369</span></span>
<span><span class="co">#&gt;  Exons in Annotation: 3823521</span></span>
<span><span class="co">#&gt;  Annotation Metadata: 6 slots</span></span></code></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tx_to_gene_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IsoformicExperiment.html">tx_to_gene</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tx_to_gene_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 6 × 2</span></span>
<span><span class="co">#&gt;   transcript_id     gene_id          </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;             &lt;chr&gt;            </span></span>
<span><span class="co">#&gt; 1 ENST00000832824.1 ENSG00000290825.2</span></span>
<span><span class="co">#&gt; 2 ENST00000832825.1 ENSG00000290825.2</span></span>
<span><span class="co">#&gt; 3 ENST00000832826.1 ENSG00000290825.2</span></span>
<span><span class="co">#&gt; 4 ENST00000832827.1 ENSG00000290825.2</span></span>
<span><span class="co">#&gt; 5 ENST00000832828.1 ENSG00000290825.2</span></span>
<span><span class="co">#&gt; 6 ENST00000832829.1 ENSG00000290825.2</span></span>
<span></span>
<span><span class="co"># rowData(se_tx) &lt;- row_data(iso)</span></span>
<span><span class="va">iso</span></span>
<span><span class="co">#&gt; &lt;IsoformicExperiment&gt;</span></span>
<span><span class="co">#&gt;  Experiment Name: 778d392d53b1394c97b69c3aec6ba177</span></span>
<span><span class="co">#&gt;  Data Path: /home/runner/.cache/R/isoformic/778d392d53b1394c97b69c3aec6ba177</span></span>
<span><span class="co">#&gt;  Assays: counts, abundance, length, infRep1, infRep2, infRep3, infRep4, infRep5, infRep6, infRep7, infRep8, infRep9, infRep10, infRep11, infRep12, infRep13, infRep14, infRep15, infRep16, infRep17, infRep18, infRep19, infRep20</span></span>
<span><span class="co">#&gt;  Samples: 6</span></span>
<span><span class="co">#&gt;  Transcripts in Assay: 111</span></span>
<span><span class="co">#&gt;  Annotation Name: gencode_v49_chr_patch_hapl_scaff_annotation_gff3</span></span>
<span><span class="co">#&gt;  Annotation Path: /home/runner/.cache/R/isoformic/gencode_v49_chr_patch_hapl_scaff_annotation_gff3</span></span>
<span><span class="co">#&gt;  Transcripts in Annotation: 533740</span></span>
<span><span class="co">#&gt;  Genes in Annotation: 86369</span></span>
<span><span class="co">#&gt;  Exons in Annotation: 3823521</span></span>
<span><span class="co">#&gt;  Annotation Metadata: 6 slots</span></span></code></pre></div>
</div>
<p>Perform transcript-level Differential Expression Analysis with fishpond.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">se_tx</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">fishpond</span><span class="fu">::</span><span class="fu"><a href="https://thelovelab.github.io/fishpond/reference/scaleInfReps.html" class="external-link">scaleInfReps</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">fishpond</span><span class="fu">::</span><span class="fu"><a href="https://thelovelab.github.io/fishpond/reference/labelKeep.html" class="external-link">labelKeep</a></span><span class="op">(</span><span class="va">y</span>, minCount <span class="op">=</span> <span class="fl">10</span>, minN <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co"># y &lt;- y[mcols(y)$keep, ]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">col_data</span><span class="op">$</span><span class="va">condition</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"treatment"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># y$condition &lt;- factor(y$condition, levels = c("control", "treatment"))</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 6 rows and 2 columns</span></span>
<span><span class="co">#&gt;                   names condition</span></span>
<span><span class="co">#&gt;             &lt;character&gt;  &lt;factor&gt;</span></span>
<span><span class="co">#&gt; SRR11498039 SRR11498039 control  </span></span>
<span><span class="co">#&gt; SRR11498040 SRR11498040 control  </span></span>
<span><span class="co">#&gt; SRR11498041 SRR11498041 control  </span></span>
<span><span class="co">#&gt; SRR11498079 SRR11498079 treatment</span></span>
<span><span class="co">#&gt; SRR11498080 SRR11498080 treatment</span></span>
<span><span class="co">#&gt; SRR11498081 SRR11498081 treatment</span></span>
<span></span>
<span><span class="co"># run Swish</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu">fishpond</span><span class="fu">::</span><span class="fu"><a href="https://thelovelab.github.io/fishpond/reference/swish.html" class="external-link">swish</a></span><span class="op">(</span><span class="va">y</span>, x <span class="op">=</span> <span class="st">"condition"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dea_tx_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">mcols</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"transcript_id"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">iso</span><span class="op">@</span><span class="va">dea</span><span class="op">$</span><span class="va">det</span> <span class="op">&lt;-</span> <span class="va">dea_tx_res</span></span>
<span></span>
<span><span class="fu"><a href="../reference/IsoformicExperiment.html">de_tx</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 111 × 9</span></span>
<span><span class="co">#&gt;    transcript_id      log10mean keep   stat log2FC pvalue locfdr qvalue det_sig</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                  &lt;dbl&gt; &lt;lgl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;chr&gt;  </span></span>
<span><span class="co">#&gt;  1 ENST00000263341.7       3.59 TRUE   3.25   6.63 0.0804  0.574  0.258 no     </span></span>
<span><span class="co">#&gt;  2 ENST00000651122.1       2.72 TRUE   3.5    4.91 0.0627  0.498  0.258 no     </span></span>
<span><span class="co">#&gt;  3 ENST00000306602.3       2.44 TRUE  -3.3   -3.75 0.0754  0.553  0.258 no     </span></span>
<span><span class="co">#&gt;  4 ENST00000392322.7       2.61 TRUE   2.45   2.62 0.179   0.912  0.466 no     </span></span>
<span><span class="co">#&gt;  5 ENST00000698141.1       2.45 TRUE   2      2.24 0.287   1      0.575 no     </span></span>
<span><span class="co">#&gt;  6 ENST00000697958.1       2.52 TRUE  -3.65  -2.01 0.0488  0.434  0.258 no     </span></span>
<span><span class="co">#&gt;  7 ENST00000673858.1       2.18 TRUE   2.1    1.98 0.266   1      0.575 no     </span></span>
<span><span class="co">#&gt;  8 ENST00000367468.10      2.92 TRUE   1.85   1.82 0.342   1      0.634 no     </span></span>
<span><span class="co">#&gt;  9 ENST00000557389.1       2.75 TRUE  -3.85  -1.67 0.0438  0.410  0.258 no     </span></span>
<span><span class="co">#&gt; 10 ENST00000330871.3       3.57 TRUE   1.5    1.63 0.423   1      0.648 no     </span></span>
<span><span class="co">#&gt; # ℹ 101 more rows</span></span></code></pre></div>
</div>
<p>Summarize Transcript SummarizedExperiment to Gene level.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_gene</span> <span class="op">&lt;-</span> <span class="fu">tximeta</span><span class="fu">::</span><span class="fu">summarizeToGene</span><span class="op">(</span></span>
<span>  <span class="va">se_tx</span>,</span>
<span>  tx2gene <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="../reference/IsoformicExperiment.html">row_data</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span>, <span class="va">transcript_id</span>, <span class="va">gene_id</span><span class="op">)</span>,</span>
<span>  skipRanges <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; summarizing abundance</span></span>
<span><span class="co">#&gt; summarizing counts</span></span>
<span><span class="co">#&gt; summarizing length</span></span>
<span><span class="co">#&gt; summarizing inferential replicates</span></span></code></pre></div>
</div>
<p>Perform Gene level Differential Expression with DESeq2.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_gene</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se_gene</span><span class="op">)</span><span class="op">$</span><span class="va">condition</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"treatment"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">se_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">se_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSet</a></span><span class="op">(</span><span class="va">se_gene</span>, design <span class="op">=</span> <span class="op">~</span><span class="va">condition</span><span class="op">)</span></span>
<span><span class="co">#&gt; using counts and average transcript lengths from tximeta</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="va">dds</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">counts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">3</span>, <span class="op">]</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co">#&gt; estimating size factors</span></span>
<span><span class="co">#&gt; using 'avgTxLength' from assays(dds), correcting for library size</span></span>
<span><span class="co">#&gt; estimating dispersions</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; fitting model and testing</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/results.html" class="external-link">results</a></span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>, <span class="st">"treatment"</span>, <span class="st">"control"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># shrink for stability on small example</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/is_installed.html" class="external-link">is_installed</a></span><span class="op">(</span><span class="st">"apeglm"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"apeglm is installed"</span><span class="op">)</span></span>
<span>  <span class="va">shrink_type</span> <span class="op">&lt;-</span> <span class="st">"apeglm"</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"apeglm is NOT installed"</span><span class="op">)</span></span>
<span>  <span class="va">shrink_type</span> <span class="op">&lt;-</span> <span class="st">"normal"</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; apeglm is installed</span></span>
<span></span>
<span><span class="va">res_shr</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/lfcShrink.html" class="external-link">lfcShrink</a></span><span class="op">(</span></span>
<span>  <span class="va">dds</span>,</span>
<span>  coef <span class="op">=</span> <span class="st">"condition_treatment_vs_control"</span>,</span>
<span>  res <span class="op">=</span> <span class="va">res</span>,</span>
<span>  type <span class="op">=</span> <span class="va">shrink_type</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; using 'apeglm' for LFC shrinkage. If used in published research, please cite:</span></span>
<span><span class="co">#&gt;     Zhu, A., Ibrahim, J.G., Love, M.I. (2018) Heavy-tailed prior distributions for</span></span>
<span><span class="co">#&gt;     sequence count data: removing the noise and preserving large differences.</span></span>
<span><span class="co">#&gt;     Bioinformatics. https://doi.org/10.1093/bioinformatics/bty895</span></span>
<span></span>
<span><span class="co"># res</span></span>
<span><span class="co"># res_shr</span></span>
<span><span class="va">dea_gene_res</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">res_shr</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"gene_id"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span></span>
<span>    log2FC <span class="op">=</span> <span class="va">log2FoldChange</span>,</span>
<span>    pvalue <span class="op">=</span> <span class="va">pvalue</span>,</span>
<span>    qvalue <span class="op">=</span> <span class="va">padj</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">iso</span><span class="op">@</span><span class="va">dea</span><span class="op">$</span><span class="va">deg</span> <span class="op">&lt;-</span> <span class="va">dea_gene_res</span></span>
<span></span>
<span><span class="fu"><a href="../reference/IsoformicExperiment.html">de_gene</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 12 × 7</span></span>
<span><span class="co">#&gt;    gene_id            baseMean       log2FC   lfcSE    pvalue   qvalue deg_sig</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                 &lt;dbl&gt;        &lt;dbl&gt;   &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;  </span></span>
<span><span class="co">#&gt;  1 ENSG00000125538.12   107.    3.09        1.47     0.000633  0.00570 yes    </span></span>
<span><span class="co">#&gt;  2 ENSG00000115415.22   464.    0.00000215  0.00144  0.432     0.668   no     </span></span>
<span><span class="co">#&gt;  3 ENSG00000271503.7  21812.   -0.00000141  0.00144  0.330     0.668   no     </span></span>
<span><span class="co">#&gt;  4 ENSG00000100906.13  1298.   -0.00000139  0.00144  0.393     0.668   no     </span></span>
<span><span class="co">#&gt;  5 ENSG00000169429.13 25857.   -0.000000744 0.00144 NA        NA       no     </span></span>
<span><span class="co">#&gt;  6 ENSG00000184557.5    254.    0.000000697 0.00144  0.680     0.765   no     </span></span>
<span><span class="co">#&gt;  7 ENSG00000169245.6     23.3  -0.000000495 0.00144  0.138     0.620   no     </span></span>
<span><span class="co">#&gt;  8 ENSG00000073756.13    63.7   0.000000483 0.00144  0.561     0.722   no     </span></span>
<span><span class="co">#&gt;  9 ENSG00000090339.10   188.   -0.000000378 0.00144  0.822     0.822   no     </span></span>
<span><span class="co">#&gt; 10 ENSG00000136244.13     6.34 -0.000000337 0.00144  0.445     0.668   no     </span></span>
<span><span class="co">#&gt; 11 ENSG00000108691.10    46.5   0.000000294 0.00144 NA        NA       no     </span></span>
<span><span class="co">#&gt; 12 ENSG00000274233.4    908.    0.000000130 0.00144 NA        NA       no</span></span></code></pre></div>
</div>
<p>Plot examples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_de_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IsoformicExperiment.html">de_gene</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">qvalue</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_log2FC.html">plot_log2fc</a></span><span class="op">(</span></span>
<span>  <span class="va">iso</span>,</span>
<span>  feature <span class="op">=</span> <span class="va">top_de_gene</span>,</span>
<span>  feature_column <span class="op">=</span> <span class="st">"gene_id"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 7 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_bar()`).</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="isoformic-preprocess_files/figure-html/unnamed-chunk-10-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>By gene name</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_to_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/IsoformicExperiment.html">row_data</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">top_de_gene</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_log2FC.html">plot_log2fc</a></span><span class="op">(</span></span>
<span>  <span class="va">iso</span>,</span>
<span>  feature <span class="op">=</span> <span class="va">gene_to_plot</span>,</span>
<span>  feature_column <span class="op">=</span> <span class="st">"gene_name"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 7 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">#&gt; (`geom_bar()`).</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="isoformic-preprocess_files/figure-html/unnamed-chunk-11-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Top DE transcripts</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/IsoformicExperiment.html">row_data</a></span><span class="op">(</span><span class="va">iso</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="st">"MTG2"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 0 × 6</span></span>
<span><span class="co">#&gt; # ℹ 6 variables: transcript_id &lt;chr&gt;, transcript_name &lt;chr&gt;,</span></span>
<span><span class="co">#&gt; #   transcript_type &lt;chr&gt;, gene_id &lt;chr&gt;, gene_name &lt;chr&gt;, gene_type &lt;chr&gt;</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_log2FC.html">plot_log2fc</a></span><span class="op">(</span></span>
<span>  <span class="va">iso</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"MTG2"</span>,</span>
<span>  feature_column <span class="op">=</span> <span class="st">"gene_name"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: No shared levels found between `names(values)` of the manual scale and the</span></span>
<span><span class="co">#&gt; data's fill values.</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="isoformic-preprocess_files/figure-html/unnamed-chunk-12-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'light-border',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config);
    }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lucio Rezende Queiroz, Lucio Rezende Queiroz, Izabela Mamede Conceicao, Luigi Marchionni, Gloria Franco.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
